{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1d3cfd4c",
   "metadata": {},
   "source": [
    "# Mapping the DropSeq data using STARsolo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c3dd8692",
   "metadata": {
    "papermill": {
     "duration": 1.864873,
     "end_time": "2022-01-27T22:27:05.313907",
     "exception": false,
     "start_time": "2022-01-27T22:27:03.449034",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "star_solo_hydrop_rna_mapping_unique_cellfilter_dropseq () {\n",
    "    local fastq_R1_filename=\"${1}\";\n",
    "    local fastq_R2_filename=\"${2}\";\n",
    "    local bam_filename=\"${3}\";\n",
    "    star_reference_dir='/lustre1/project/stg_00002/lcb/fderop/data/00000000_genomes/mm10_STAR_2.7.8_premrna';\n",
    "    sjdbgtf='/lustre1/project/stg_00002/lcb/fderop/data/00000000_genomes/mm10_STAR_2.7.8_premrna/genes.gtf'\n",
    "    #module load STAR/2.7.5b-foss-2018a;\n",
    "    #module load SAMtools/1.10-foss-2018a;\n",
    "    # Map with STAR solo\n",
    "    /lustre1/project/stg_00002/lcb/fderop/scripts/STAR-2.7.8a/bin/Linux_x86_64/STAR \\\n",
    "        --runThreadN 36 \\\n",
    "        --runMode alignReads \\\n",
    "        --outSAMtype BAM SortedByCoordinate \\\n",
    "        --sysShell /bin/bash \\\n",
    "        --genomeDir \"${star_reference_dir}\" \\\n",
    "        --readFilesIn \"${fastq_R1_filename}\" \"${fastq_R2_filename}\" \\\n",
    "        --readFilesCommand 'gzip -c -d' \\\n",
    "        --soloCBwhitelist None \\\n",
    "        --soloType CB_UMI_Simple \\\n",
    "        --soloCBstart  1 \\\n",
    "        --soloCBlen 12 \\\n",
    "        --soloUMIstart 13 \\\n",
    "        --soloUMIlen 8\\\n",
    "        --sjdbGTFfile $sjdbgtf \\\n",
    "        --soloCellFilter CellRanger2.2 2000 0.99 10 \\\n",
    "        --outFilterMultimapNmax 1 \\\n",
    "        --outSAMattributes NH HI AS nM CB UB CR CY UR UY \\\n",
    "        --outFileNamePrefix ${bam_filename%bam} \\\n",
    "        --outReadsUnmapped Fastx \\\n",
    "        --quantMode GeneCounts \\\n",
    "        --bamRemoveDuplicatesType UniqueIdentical \\\n",
    "        --soloFeatures Gene GeneFull\n",
    "        # Index BAM file.\n",
    "    # samtools index \"${bam_filename%bam}Aligned.sortedByCoord.out.bam\"\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d1d6a69",
   "metadata": {
    "papermill": {
     "duration": 0.008944,
     "end_time": "2022-01-27T22:27:05.363011",
     "exception": false,
     "start_time": "2022-01-27T22:27:05.354067",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Now we run this star solo script in a for loop on all the R2 fastq files in `fastq/`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e072a703",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-01-27T22:27:05.373643Z",
     "iopub.status.busy": "2022-01-27T22:27:05.373250Z",
     "iopub.status.idle": "2022-01-28T04:10:06.014436Z",
     "shell.execute_reply": "2022-01-28T04:10:05.951866Z"
    },
    "papermill": {
     "duration": 20580.711883,
     "end_time": "2022-01-28T04:10:06.078631",
     "exception": true,
     "start_time": "2022-01-27T22:27:05.366748",
     "status": "failed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘demultiplexed’: File exists\n",
      "fastq_dropseq/SRR1853178\n",
      "Jan 27 23:27:06 ..... started STAR run\n",
      "Jan 27 23:27:06 ..... loading genome\n",
      "Jan 27 23:27:22 ..... processing annotations GTF\n",
      "Jan 27 23:27:23 ..... started mapping\n",
      "Jan 28 00:08:03 ..... finished mapping\n",
      "Jan 28 00:08:08 ..... started Solo counting\n",
      "Jan 28 00:19:21 ..... finished Solo counting\n",
      "Jan 28 00:19:21 ..... started sorting BAM\n",
      "Jan 28 00:27:47 ..... finished successfully\n",
      "fastq_dropseq/SRR1853179\n",
      "Jan 28 00:28:00 ..... started STAR run\n",
      "Jan 28 00:28:00 ..... loading genome\n",
      "Jan 28 00:28:34 ..... processing annotations GTF\n",
      "Jan 28 00:28:40 ..... started mapping\n",
      "Jan 28 01:10:24 ..... finished mapping\n",
      "Jan 28 01:10:30 ..... started Solo counting\n",
      "Jan 28 01:22:25 ..... finished Solo counting\n",
      "Jan 28 01:22:25 ..... started sorting BAM\n",
      "Jan 28 01:30:54 ..... finished successfully\n",
      "fastq_dropseq/SRR1853180\n",
      "Jan 28 01:31:10 ..... started STAR run\n",
      "Jan 28 01:31:10 ..... loading genome\n",
      "Jan 28 01:31:42 ..... processing annotations GTF\n",
      "Jan 28 01:31:47 ..... started mapping\n",
      "Jan 28 02:04:48 ..... finished mapping\n",
      "Jan 28 02:04:53 ..... started Solo counting\n",
      "Jan 28 02:13:14 ..... finished Solo counting\n",
      "Jan 28 02:13:14 ..... started sorting BAM\n",
      "Jan 28 02:19:39 ..... finished successfully\n",
      "fastq_dropseq/SRR1853181\n",
      "Jan 28 02:19:59 ..... started STAR run\n",
      "Jan 28 02:19:59 ..... loading genome\n",
      "Jan 28 02:20:32 ..... processing annotations GTF\n",
      "Jan 28 02:20:38 ..... started mapping\n",
      "Jan 28 03:04:26 ..... finished mapping\n",
      "Jan 28 03:04:31 ..... started Solo counting\n",
      "Jan 28 03:12:43 ..... finished Solo counting\n",
      "Jan 28 03:12:43 ..... started sorting BAM\n",
      "Jan 28 03:17:51 ..... finished successfully\n",
      "fastq_dropseq/SRR1853182\n",
      "Jan 28 03:18:08 ..... started STAR run\n",
      "Jan 28 03:18:08 ..... loading genome\n",
      "Jan 28 03:18:41 ..... processing annotations GTF\n",
      "Jan 28 03:18:47 ..... started mapping\n",
      "Jan 28 03:58:18 ..... finished mapping\n",
      "Jan 28 03:58:23 ..... started Solo counting\n",
      "Jan 28 04:08:39 ..... finished Solo counting\n",
      "Jan 28 04:08:39 ..... started sorting BAM\n",
      "Jan 28 04:16:47 ..... finished successfully\n",
      "fastq_dropseq/SRR1853183\n",
      "Jan 28 04:16:59 ..... started STAR run\n",
      "Jan 28 04:16:59 ..... loading genome\n",
      "Jan 28 04:17:35 ..... processing annotations GTF\n",
      "Jan 28 04:17:42 ..... started mapping\n",
      "Jan 28 04:52:49 ..... finished mapping\n",
      "Jan 28 04:52:55 ..... started Solo counting\n",
      "Jan 28 05:02:01 ..... finished Solo counting\n",
      "Jan 28 05:02:01 ..... started sorting BAM\n",
      "Jan 28 05:08:59 ..... finished successfully\n",
      "fastq_dropseq/SRR1853184\n",
      "Jan 28 05:09:06 ..... started STAR run\n",
      "Jan 28 05:09:07 ..... loading genome\n",
      "Jan 28 05:09:40 ..... processing annotations GTF\n",
      "Jan 28 05:09:47 ..... started mapping\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "Bus error\n"
     ]
    },
    {
     "ename": "",
     "evalue": "135",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "dir=demultiplexed_dropseq\n",
    "mkdir $dir\n",
    "for sample in fastq_dropseq/*_2.fastq.gz\n",
    "do\n",
    "    NAME=${sample%_2.fastq.gz}\n",
    "    OUT=${NAME##*/}\n",
    "    echo $NAME\n",
    "    star_solo_hydrop_rna_mapping_unique_cellfilter_dropseq \\\n",
    "        ${NAME}_2.fastq.gz \\\n",
    "        ${NAME}_1.fastq.gz \\\n",
    "        $dir/$OUT.bam\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "21d845a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘demultiplexed_dropseq’: File exists\n",
      "fastq_dropseq/SRR1853184\n",
      "Jan 28 11:32:30 ..... started STAR run\n",
      "Jan 28 11:32:31 ..... loading genome\n",
      "Jan 28 11:33:00 ..... processing annotations GTF\n",
      "Jan 28 11:33:07 ..... started mapping\n",
      "Jan 28 12:13:02 ..... finished mapping\n",
      "Jan 28 12:13:05 ..... started Solo counting\n",
      "Jan 28 12:23:56 ..... finished Solo counting\n",
      "Jan 28 12:23:56 ..... started sorting BAM\n",
      "Jan 28 12:31:58 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "dir=demultiplexed_dropseq\n",
    "mkdir $dir\n",
    "for sample in fastq_dropseq/SRR1853184_2.fastq.gz\n",
    "do\n",
    "    NAME=${sample%_2.fastq.gz}\n",
    "    OUT=${NAME##*/}\n",
    "    echo $NAME\n",
    "    star_solo_hydrop_rna_mapping_unique_cellfilter_dropseq \\\n",
    "        ${NAME}_2.fastq.gz \\\n",
    "        ${NAME}_1.fastq.gz \\\n",
    "        $dir/$OUT.bam\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e11dfad",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "source": [
    "Now, we can get a summary of the statistics below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fab51657",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR1853178.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,510210716\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.757257\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.62094\n",
      "Reads Mapped to Genome: Unique+Multiple,0.545704\n",
      "Reads Mapped to Genome: Unique,0.545704\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.372239\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.362322\n",
      "Estimated Number of Cells,2629\n",
      "Reads in Cells Mapped to Unique Genes,54743780\n",
      "Fraction of Reads in Cells,0.296135\n",
      "Mean Reads per Cell,20823\n",
      "Median Reads per Cell,15973\n",
      "UMIs in Cells,7773057\n",
      "Mean UMI per Cell,2956\n",
      "Median UMI per Cell,2202\n",
      "Mean Genes per Cell,1651\n",
      "Median Genes per Cell,1391\n",
      "Total Genes Detected,22048\n",
      "\n",
      "SRR1853179.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,524537760\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.768773\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.620952\n",
      "Reads Mapped to Genome: Unique+Multiple,0.543425\n",
      "Reads Mapped to Genome: Unique,0.543425\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.376026\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.365333\n",
      "Estimated Number of Cells,2208\n",
      "Reads in Cells Mapped to Unique Genes,57388174\n",
      "Fraction of Reads in Cells,0.299472\n",
      "Mean Reads per Cell,25991\n",
      "Median Reads per Cell,19780\n",
      "UMIs in Cells,7713453\n",
      "Mean UMI per Cell,3493\n",
      "Median UMI per Cell,2625\n",
      "Mean Genes per Cell,1863\n",
      "Median Genes per Cell,1603\n",
      "Total Genes Detected,22163\n",
      "\n",
      "SRR1853180.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,411365100\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.830703\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.566498\n",
      "Reads Mapped to Genome: Unique+Multiple,0.542113\n",
      "Reads Mapped to Genome: Unique,0.542113\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.363961\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.353768\n",
      "Estimated Number of Cells,1697\n",
      "Reads in Cells Mapped to Unique Genes,52238936\n",
      "Fraction of Reads in Cells,0.358962\n",
      "Mean Reads per Cell,30783\n",
      "Median Reads per Cell,22655\n",
      "UMIs in Cells,4566499\n",
      "Mean UMI per Cell,2690\n",
      "Median UMI per Cell,1953\n",
      "Mean Genes per Cell,1527\n",
      "Median Genes per Cell,1266\n",
      "Total Genes Detected,21708\n",
      "\n",
      "SRR1853181.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,486748074\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.674636\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.608495\n",
      "Reads Mapped to Genome: Unique+Multiple,0.332314\n",
      "Reads Mapped to Genome: Unique,0.332314\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.239943\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.23384\n",
      "Estimated Number of Cells,1798\n",
      "Reads in Cells Mapped to Unique Genes,38185717\n",
      "Fraction of Reads in Cells,0.335488\n",
      "Mean Reads per Cell,21237\n",
      "Median Reads per Cell,15512\n",
      "UMIs in Cells,8390763\n",
      "Mean UMI per Cell,4666\n",
      "Median UMI per Cell,3400\n",
      "Mean Genes per Cell,2211\n",
      "Median Genes per Cell,1907\n",
      "Total Genes Detected,22145\n",
      "\n",
      "SRR1853182.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,491102701\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.771998\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.637598\n",
      "Reads Mapped to Genome: Unique+Multiple,0.535425\n",
      "Reads Mapped to Genome: Unique,0.535425\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.382315\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.372148\n",
      "Estimated Number of Cells,2221\n",
      "Reads in Cells Mapped to Unique Genes,67329155\n",
      "Fraction of Reads in Cells,0.368397\n",
      "Mean Reads per Cell,30314\n",
      "Median Reads per Cell,23341\n",
      "UMIs in Cells,10082819\n",
      "Mean UMI per Cell,4539\n",
      "Median UMI per Cell,3363\n",
      "Mean Genes per Cell,2182\n",
      "Median Genes per Cell,1872\n",
      "Total Genes Detected,22631\n",
      "\n",
      "SRR1853183.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,438843919\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.844932\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.591365\n",
      "Reads Mapped to Genome: Unique+Multiple,0.551568\n",
      "Reads Mapped to Genome: Unique,0.551568\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.435267\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.423811\n",
      "Estimated Number of Cells,2249\n",
      "Reads in Cells Mapped to Unique Genes,79424580\n",
      "Fraction of Reads in Cells,0.427044\n",
      "Mean Reads per Cell,35315\n",
      "Median Reads per Cell,26822\n",
      "UMIs in Cells,6949914\n",
      "Mean UMI per Cell,3090\n",
      "Median UMI per Cell,2272\n",
      "Mean Genes per Cell,1716\n",
      "Median Genes per Cell,1447\n",
      "Total Genes Detected,21949\n",
      "\n",
      "SRR1853184.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,494030802\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.725467\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.600765\n",
      "Reads Mapped to Genome: Unique+Multiple,0.562359\n",
      "Reads Mapped to Genome: Unique,0.562359\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.416486\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.405025\n",
      "Estimated Number of Cells,1953\n",
      "Reads in Cells Mapped to Unique Genes,45691499\n",
      "Fraction of Reads in Cells,0.228349\n",
      "Mean Reads per Cell,23395\n",
      "Median Reads per Cell,16929\n",
      "UMIs in Cells,8016444\n",
      "Mean UMI per Cell,4104\n",
      "Median UMI per Cell,2916\n",
      "Mean Genes per Cell,2005\n",
      "Median Genes per Cell,1688\n",
      "Total Genes Detected,21366\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for sample in demultiplexed_dropseq/*/Gene/Summary.csv\n",
    "do\n",
    "    echo ${sample#*/}\n",
    "    cat $sample\n",
    "    cut -d, -f2 $sample > ${sample%/Gene/Summary.csv}/Gene/Summary_numbers.csv\n",
    "    printf \"\\n\"\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 20587.968293,
   "end_time": "2022-01-28T04:10:07.788170",
   "environment_variables": {},
   "exception": true,
   "input_path": "3_mapping_dropseq.ipynb",
   "output_path": "3_mapping_dropseq.OUT.ipynb",
   "parameters": {},
   "start_time": "2022-01-27T22:26:59.819877",
   "version": "2.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
