{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# first, generate a barcode read from read 2, 4, 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "basically followed this: https://github.com/alexdobin/STAR/issues/825"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "module load mawk\n",
    "for file in fastq_indrop/*_1.fastq.gz\n",
    "do\n",
    "    echo $file\n",
    "    sample=`basename ${file%_1.fastq.gz}`\n",
    "    echo $sample\n",
    "    SAMPLE_R2=fastq_indrop/${sample}_2.fastq.gz\n",
    "    SAMPLE_R4=fastq_indrop/${sample}_4.fastq.gz\n",
    "    SAMPLE_R24=fastq_indrop/${sample}_24.fastq.gz\n",
    "    echo $SAMPLE_R24\n",
    "    paste <(zcat ${SAMPLE_R2}) <(zcat ${SAMPLE_R4}) | mawk '{if (NR%4==1) {print $1 \" 2:4:paste\"} else if (NR%4==3) {print $1} else {print $1 $2}}' | gzip > ${SAMPLE_R24} &\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "jobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "paste \\ \n",
    "    <(zcat ${SAMPLE_R2}) \\\n",
    "    <(zcat ${SAMPLE_R3}) \\\n",
    "    <(zcat ${SAMPLE_R4}) \\\n",
    "    | mawk '{if ((NR%4==1) && ($1 == $3) && ($3 == $5)) {print $1 \" 2:3:4:paste\"} else if ((NR%4==1) && ($1 != $3)) {print \"singleton\"} else if ((NR%4==1) && ($3 != $5)) {print \"singleton\"} else if (NR%4==3) {print $1} else {print $1 $2 $3}}' \\\n",
    "    | gzip > ${SAMPLE_R234.fastq.gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# then, generate bc list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "module load mawk\n",
    "cartesian(){\n",
    "awk -F '\\t' \\\n",
    "    -v \"bc1_list=$1\" \\\n",
    "    -v \"bc2_list=$2\" '\n",
    "    {\n",
    "        if ( FILENAME == bc1_list ) {\n",
    "            BC1[$0];\n",
    "        }\n",
    "        else if ( FILENAME == bc2_list ) {\n",
    "             BC2[$0];\n",
    "        }\n",
    "    } END {\n",
    "        for (bc1 in BC1) {\n",
    "           for (bc2 in BC2) {\n",
    "               print bc1 bc2\n",
    "           }\n",
    "        }\n",
    "     } ' $1 $2 > $3\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat 20220127_indrop_barcodes_kalish_384x384/gel_barcode2_list.txt | tr 'ACGT' 'TGCA' | rev > 20220127_indrop_barcodes_kalish_384x384/gel_barcode2_list_revcomp.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cartesian 20220127_indrop_barcodes_kalish_384x384/gel_barcode2_list.txt 20220127_indrop_barcodes_kalish_384x384/gel_barcode2_list_revcomp.txt 20220127_indrop_barcodes_kalish_384x384/full_whitelist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# then map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pos 1-8: bc1  \n",
    "pos 9-16: bc2 (reverse complement of whitelist)  \n",
    "pos 17-22: UMI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "star_solo_indrop_rna_mapping_unique_cellfilter () {\n",
    "    local fastq_R1_filename=\"${1}\";\n",
    "    local fastq_R2_filename=\"${2}\";\n",
    "    local bam_filename=\"${3}\";\n",
    "    # 3 barcode lists:\n",
    "    local whitelist_part1_filename='/lustre1/project/stg_00002/lcb/fderop/data/00000000_barcodes/20220127_indrop_barcodes_kalish_384x384/gel_barcode2_list.txt';\n",
    "    local whitelist_part2_filename='/lustre1/project/stg_00002/lcb/fderop/data/00000000_barcodes/20220127_indrop_barcodes_kalish_384x384/gel_barcode2_list_revcomp.txt';\n",
    "    star_reference_dir='/lustre1/project/stg_00002/lcb/fderop/data/00000000_genomes/mm10_STAR_2.7.8_premrna';\n",
    "    sjdbgtf='/lustre1/project/stg_00002/lcb/fderop/data/00000000_genomes/mm10_STAR_2.7.8_premrna/genes.gtf'\n",
    "    #module load STAR/2.7.5b-foss-2018a;\n",
    "    #module load SAMtools/1.10-foss-2018a;\n",
    "    # Map with STAR solo\n",
    "    /lustre1/project/stg_00002/lcb/fderop/scripts/STAR-2.7.8a/bin/Linux_x86_64/STAR \\\n",
    "        --runThreadN 36 \\\n",
    "        --runMode alignReads \\\n",
    "        --outSAMtype BAM SortedByCoordinate \\\n",
    "        --sysShell /bin/bash \\\n",
    "        --genomeDir \"${star_reference_dir}\" \\\n",
    "        --readFilesIn \"${fastq_R1_filename}\" \"${fastq_R2_filename}\" \\\n",
    "        --readFilesCommand 'gzip -c -d' \\\n",
    "        --soloCBwhitelist \"${whitelist_part1_filename}\" \"${whitelist_part2_filename}\" \\\n",
    "        --soloType CB_UMI_Complex \\\n",
    "        --soloCBposition 0_0_0_7 0_8_0_15 \\\n",
    "        --soloUMIposition 0_16_0_21 \\\n",
    "        --sjdbGTFfile $sjdbgtf \\\n",
    "        --soloCellFilter CellRanger2.2 2000 0.99 10 \\\n",
    "        --soloCBmatchWLtype 1MM \\\n",
    "        --outFilterMultimapNmax 1 \\\n",
    "        --outSAMattributes NH HI AS nM CB UB CR CY UR UY \\\n",
    "        --outFileNamePrefix ${bam_filename%bam} \\\n",
    "        --outReadsUnmapped Fastx \\\n",
    "        --quantMode GeneCounts \\\n",
    "        --bamRemoveDuplicatesType UniqueIdentical \\\n",
    "        --soloFeatures Gene GeneFull\n",
    "        # Index BAM file.\n",
    "    # samtools index \"${bam_filename%bam}Aligned.sortedByCoord.out.bam\"\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we run this star solo script in a for loop on all the R2 fastq files in `fastq/`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘demultiplexed_indrop’: File exists\n",
      "fastq_indrop/SRR10545068\n",
      "Jan 28 16:03:41 ..... started STAR run\n",
      "Jan 28 16:03:41 ..... loading genome\n",
      "Jan 28 16:04:09 ..... processing annotations GTF\n",
      "Jan 28 16:04:15 ..... started mapping\n",
      "Jan 28 16:14:44 ..... finished mapping\n",
      "Jan 28 16:14:47 ..... started Solo counting\n",
      "Jan 28 16:19:06 ..... finished Solo counting\n",
      "Jan 28 16:19:06 ..... started sorting BAM\n",
      "Jan 28 16:20:43 ..... finished successfully\n",
      "fastq_indrop/SRR10545069\n",
      "Jan 28 16:20:53 ..... started STAR run\n",
      "Jan 28 16:20:53 ..... loading genome\n",
      "Jan 28 16:21:11 ..... processing annotations GTF\n",
      "Jan 28 16:21:12 ..... started mapping\n",
      "Jan 28 16:29:11 ..... finished mapping\n",
      "Jan 28 16:29:15 ..... started Solo counting\n",
      "Jan 28 16:30:12 ..... finished Solo counting\n",
      "Jan 28 16:30:12 ..... started sorting BAM\n",
      "Jan 28 16:31:22 ..... finished successfully\n",
      "fastq_indrop/SRR10545070\n",
      "Jan 28 16:31:29 ..... started STAR run\n",
      "Jan 28 16:31:29 ..... loading genome\n",
      "Jan 28 16:31:44 ..... processing annotations GTF\n",
      "Jan 28 16:31:45 ..... started mapping\n",
      "Jan 28 16:39:12 ..... finished mapping\n",
      "Jan 28 16:39:15 ..... started Solo counting\n",
      "Jan 28 16:40:03 ..... finished Solo counting\n",
      "Jan 28 16:40:04 ..... started sorting BAM\n",
      "Jan 28 16:41:10 ..... finished successfully\n",
      "fastq_indrop/SRR10545071\n",
      "Jan 28 16:41:19 ..... started STAR run\n",
      "Jan 28 16:41:19 ..... loading genome\n",
      "Jan 28 16:41:32 ..... processing annotations GTF\n",
      "Jan 28 16:41:34 ..... started mapping\n",
      "Jan 28 16:49:20 ..... finished mapping\n",
      "Jan 28 16:49:23 ..... started Solo counting\n",
      "Jan 28 16:50:24 ..... finished Solo counting\n",
      "Jan 28 16:50:24 ..... started sorting BAM\n",
      "Jan 28 16:51:34 ..... finished successfully\n",
      "fastq_indrop/SRR10545072\n",
      "Jan 28 16:51:43 ..... started STAR run\n",
      "Jan 28 16:51:43 ..... loading genome\n",
      "Jan 28 16:51:57 ..... processing annotations GTF\n",
      "Jan 28 16:51:58 ..... started mapping\n",
      "Jan 28 17:01:56 ..... finished mapping\n",
      "Jan 28 17:01:59 ..... started Solo counting\n",
      "Jan 28 17:04:15 ..... finished Solo counting\n",
      "Jan 28 17:04:15 ..... started sorting BAM\n",
      "Jan 28 17:05:45 ..... finished successfully\n",
      "fastq_indrop/SRR10545073\n",
      "Jan 28 17:05:57 ..... started STAR run\n",
      "Jan 28 17:05:57 ..... loading genome\n",
      "Jan 28 17:06:26 ..... processing annotations GTF\n",
      "Jan 28 17:06:32 ..... started mapping\n",
      "Jan 28 17:16:06 ..... finished mapping\n",
      "Jan 28 17:16:09 ..... started Solo counting\n",
      "Jan 28 17:18:12 ..... finished Solo counting\n",
      "Jan 28 17:18:12 ..... started sorting BAM\n",
      "Jan 28 17:19:38 ..... finished successfully\n",
      "fastq_indrop/SRR10545074\n",
      "Jan 28 17:19:49 ..... started STAR run\n",
      "Jan 28 17:19:49 ..... loading genome\n",
      "Jan 28 17:20:18 ..... processing annotations GTF\n",
      "Jan 28 17:20:23 ..... started mapping\n",
      "Jan 28 17:30:13 ..... finished mapping\n",
      "Jan 28 17:30:15 ..... started Solo counting\n",
      "Jan 28 17:32:15 ..... finished Solo counting\n",
      "Jan 28 17:32:15 ..... started sorting BAM\n",
      "Jan 28 17:33:43 ..... finished successfully\n",
      "fastq_indrop/SRR10545075\n",
      "Jan 28 17:33:54 ..... started STAR run\n",
      "Jan 28 17:33:54 ..... loading genome\n",
      "Jan 28 17:34:25 ..... processing annotations GTF\n",
      "Jan 28 17:34:31 ..... started mapping\n",
      "Jan 28 17:43:41 ..... finished mapping\n",
      "Jan 28 17:43:44 ..... started Solo counting\n",
      "Jan 28 17:44:51 ..... finished Solo counting\n",
      "Jan 28 17:44:51 ..... started sorting BAM\n",
      "Jan 28 17:46:18 ..... finished successfully\n",
      "fastq_indrop/SRR10545076\n",
      "Jan 28 17:46:30 ..... started STAR run\n",
      "Jan 28 17:46:30 ..... loading genome\n",
      "Jan 28 17:47:00 ..... processing annotations GTF\n",
      "Jan 28 17:47:01 ..... started mapping\n",
      "Jan 28 17:56:21 ..... finished mapping\n",
      "Jan 28 17:56:23 ..... started Solo counting\n",
      "Jan 28 17:59:05 ..... finished Solo counting\n",
      "Jan 28 17:59:05 ..... started sorting BAM\n",
      "Jan 28 18:00:36 ..... finished successfully\n",
      "fastq_indrop/SRR10545077\n",
      "Jan 28 18:00:46 ..... started STAR run\n",
      "Jan 28 18:00:46 ..... loading genome\n",
      "Jan 28 18:01:01 ..... processing annotations GTF\n",
      "Jan 28 18:01:07 ..... started mapping\n",
      "Jan 28 18:10:17 ..... finished mapping\n",
      "Jan 28 18:10:20 ..... started Solo counting\n",
      "Jan 28 18:12:47 ..... finished Solo counting\n",
      "Jan 28 18:12:47 ..... started sorting BAM\n",
      "Jan 28 18:14:22 ..... finished successfully\n",
      "fastq_indrop/SRR10545078\n",
      "Jan 28 18:14:30 ..... started STAR run\n",
      "Jan 28 18:14:30 ..... loading genome\n",
      "Jan 28 18:14:44 ..... processing annotations GTF\n",
      "Jan 28 18:14:46 ..... started mapping\n",
      "Jan 28 18:23:59 ..... finished mapping\n",
      "Jan 28 18:24:02 ..... started Solo counting\n",
      "Jan 28 18:25:30 ..... finished Solo counting\n",
      "Jan 28 18:25:30 ..... started sorting BAM\n",
      "Jan 28 18:26:58 ..... finished successfully\n",
      "fastq_indrop/SRR10545079\n",
      "Jan 28 18:27:11 ..... started STAR run\n",
      "Jan 28 18:27:11 ..... loading genome\n",
      "Jan 28 18:27:42 ..... processing annotations GTF\n",
      "Jan 28 18:27:48 ..... started mapping\n",
      "Jan 28 18:35:32 ..... finished mapping\n",
      "Jan 28 18:35:35 ..... started Solo counting\n",
      "Jan 28 18:36:29 ..... finished Solo counting\n",
      "Jan 28 18:36:29 ..... started sorting BAM\n",
      "Jan 28 18:37:41 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "dir=demultiplexed_indrop\n",
    "mkdir $dir\n",
    "for sample in fastq_indrop/*_1.fastq.gz\n",
    "do\n",
    "    NAME=${sample%_1.fastq.gz}\n",
    "    OUT=${NAME##*/}\n",
    "    echo $NAME\n",
    "    star_solo_indrop_rna_mapping_unique_cellfilter \\\n",
    "        ${NAME}_1.fastq.gz \\\n",
    "        ${NAME}_24.fastq.gz \\\n",
    "        $dir/$OUT.bam\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we can get a summary of the statistics below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR10545068.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,136252173\n",
      "Reads With Valid Barcodes,0.440879\n",
      "Sequencing Saturation,0.266368\n",
      "Q30 Bases in CB+UMI,0.832998\n",
      "Q30 Bases in RNA read,0.800453\n",
      "Reads Mapped to Genome: Unique+Multiple,0.488978\n",
      "Reads Mapped to Genome: Unique,0.488978\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.301477\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.293621\n",
      "Estimated Number of Cells,14363\n",
      "Reads in Cells Mapped to Unique Genes,24349722\n",
      "Fraction of Reads in Cells,0.608644\n",
      "Mean Reads per Cell,1695\n",
      "Median Reads per Cell,1432\n",
      "UMIs in Cells,17563343\n",
      "Mean UMI per Cell,1222\n",
      "Median UMI per Cell,1044\n",
      "Mean Genes per Cell,896\n",
      "Median Genes per Cell,806\n",
      "Total Genes Detected,21125\n",
      "\n",
      "SRR10545069.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,103036640\n",
      "Reads With Valid Barcodes,0.381989\n",
      "Sequencing Saturation,0.196112\n",
      "Q30 Bases in CB+UMI,0.732578\n",
      "Q30 Bases in RNA read,0.711272\n",
      "Reads Mapped to Genome: Unique+Multiple,0.450167\n",
      "Reads Mapped to Genome: Unique,0.450167\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.249958\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.243351\n",
      "Estimated Number of Cells,12538\n",
      "Reads in Cells Mapped to Unique Genes,14228154\n",
      "Fraction of Reads in Cells,0.567445\n",
      "Mean Reads per Cell,1134\n",
      "Median Reads per Cell,945\n",
      "UMIs in Cells,11255164\n",
      "Mean UMI per Cell,897\n",
      "Median UMI per Cell,755\n",
      "Mean Genes per Cell,692\n",
      "Median Genes per Cell,612\n",
      "Total Genes Detected,20384\n",
      "\n",
      "SRR10545070.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,98859416\n",
      "Reads With Valid Barcodes,0.408675\n",
      "Sequencing Saturation,0.196926\n",
      "Q30 Bases in CB+UMI,0.746938\n",
      "Q30 Bases in RNA read,0.710311\n",
      "Reads Mapped to Genome: Unique+Multiple,0.447653\n",
      "Reads Mapped to Genome: Unique,0.447653\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.264455\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.257502\n",
      "Estimated Number of Cells,12668\n",
      "Reads in Cells Mapped to Unique Genes,14533296\n",
      "Fraction of Reads in Cells,0.570907\n",
      "Mean Reads per Cell,1147\n",
      "Median Reads per Cell,955\n",
      "UMIs in Cells,11493802\n",
      "Mean UMI per Cell,907\n",
      "Median UMI per Cell,764\n",
      "Mean Genes per Cell,699\n",
      "Median Genes per Cell,619\n",
      "Total Genes Detected,20493\n",
      "\n",
      "SRR10545071.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,100660676\n",
      "Reads With Valid Barcodes,0.391889\n",
      "Sequencing Saturation,0.195025\n",
      "Q30 Bases in CB+UMI,0.739465\n",
      "Q30 Bases in RNA read,0.712597\n",
      "Reads Mapped to Genome: Unique+Multiple,0.452498\n",
      "Reads Mapped to Genome: Unique,0.452498\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.255742\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.248998\n",
      "Estimated Number of Cells,12618\n",
      "Reads in Cells Mapped to Unique Genes,14276253\n",
      "Fraction of Reads in Cells,0.569584\n",
      "Mean Reads per Cell,1131\n",
      "Median Reads per Cell,942\n",
      "UMIs in Cells,11315094\n",
      "Mean UMI per Cell,896\n",
      "Median UMI per Cell,756\n",
      "Mean Genes per Cell,692\n",
      "Median Genes per Cell,612\n",
      "Total Genes Detected,20449\n",
      "\n",
      "SRR10545072.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,134789164\n",
      "Reads With Valid Barcodes,0.416603\n",
      "Sequencing Saturation,0.255584\n",
      "Q30 Bases in CB+UMI,0.817862\n",
      "Q30 Bases in RNA read,0.798391\n",
      "Reads Mapped to Genome: Unique+Multiple,0.483215\n",
      "Reads Mapped to Genome: Unique,0.483215\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.284974\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.277531\n",
      "Estimated Number of Cells,14249\n",
      "Reads in Cells Mapped to Unique Genes,22663375\n",
      "Fraction of Reads in Cells,0.605839\n",
      "Mean Reads per Cell,1590\n",
      "Median Reads per Cell,1343\n",
      "UMIs in Cells,16585294\n",
      "Mean UMI per Cell,1163\n",
      "Median UMI per Cell,992\n",
      "Mean Genes per Cell,861\n",
      "Median Genes per Cell,775\n",
      "Total Genes Detected,21013\n",
      "\n",
      "SRR10545073.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,130210119\n",
      "Reads With Valid Barcodes,0.439545\n",
      "Sequencing Saturation,0.25687\n",
      "Q30 Bases in CB+UMI,0.82782\n",
      "Q30 Bases in RNA read,0.791933\n",
      "Reads Mapped to Genome: Unique+Multiple,0.48634\n",
      "Reads Mapped to Genome: Unique,0.48634\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.299095\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.29128\n",
      "Estimated Number of Cells,14411\n",
      "Reads in Cells Mapped to Unique Genes,23105678\n",
      "Fraction of Reads in Cells,0.609205\n",
      "Mean Reads per Cell,1603\n",
      "Median Reads per Cell,1354\n",
      "UMIs in Cells,16891716\n",
      "Mean UMI per Cell,1172\n",
      "Median UMI per Cell,1001\n",
      "Mean Genes per Cell,866\n",
      "Median Genes per Cell,779\n",
      "Total Genes Detected,21100\n",
      "\n",
      "SRR10545074.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,131473567\n",
      "Reads With Valid Barcodes,0.422575\n",
      "Sequencing Saturation,0.253033\n",
      "Q30 Bases in CB+UMI,0.821877\n",
      "Q30 Bases in RNA read,0.796305\n",
      "Reads Mapped to Genome: Unique+Multiple,0.485901\n",
      "Reads Mapped to Genome: Unique,0.485901\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.288511\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.280963\n",
      "Estimated Number of Cells,14248\n",
      "Reads in Cells Mapped to Unique Genes,22386739\n",
      "Fraction of Reads in Cells,0.606043\n",
      "Mean Reads per Cell,1571\n",
      "Median Reads per Cell,1327\n",
      "UMIs in Cells,16443474\n",
      "Mean UMI per Cell,1154\n",
      "Median UMI per Cell,985\n",
      "Mean Genes per Cell,855\n",
      "Median Genes per Cell,767\n",
      "Total Genes Detected,20962\n",
      "\n",
      "SRR10545075.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,125843005\n",
      "Reads With Valid Barcodes,0.437122\n",
      "Sequencing Saturation,0.259587\n",
      "Q30 Bases in CB+UMI,0.801609\n",
      "Q30 Bases in RNA read,0.767215\n",
      "Reads Mapped to Genome: Unique+Multiple,0.496859\n",
      "Reads Mapped to Genome: Unique,0.496859\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.300357\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.292619\n",
      "Estimated Number of Cells,13365\n",
      "Reads in Cells Mapped to Unique Genes,21844268\n",
      "Fraction of Reads in Cells,0.593206\n",
      "Mean Reads per Cell,1634\n",
      "Median Reads per Cell,1363\n",
      "UMIs in Cells,15840113\n",
      "Mean UMI per Cell,1185\n",
      "Median UMI per Cell,1006\n",
      "Mean Genes per Cell,872\n",
      "Median Genes per Cell,781\n",
      "Total Genes Detected,21070\n",
      "\n",
      "SRR10545076.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,125597295\n",
      "Reads With Valid Barcodes,0.413883\n",
      "Sequencing Saturation,0.25111\n",
      "Q30 Bases in CB+UMI,0.792616\n",
      "Q30 Bases in RNA read,0.773506\n",
      "Reads Mapped to Genome: Unique+Multiple,0.492923\n",
      "Reads Mapped to Genome: Unique,0.492923\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.284902\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.277517\n",
      "Estimated Number of Cells,13375\n",
      "Reads in Cells Mapped to Unique Genes,20680261\n",
      "Fraction of Reads in Cells,0.593316\n",
      "Mean Reads per Cell,1546\n",
      "Median Reads per Cell,1292\n",
      "UMIs in Cells,15173551\n",
      "Mean UMI per Cell,1134\n",
      "Median UMI per Cell,966\n",
      "Mean Genes per Cell,841\n",
      "Median Genes per Cell,754\n",
      "Total Genes Detected,20931\n",
      "\n",
      "SRR10545077.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,120173234\n",
      "Reads With Valid Barcodes,0.434949\n",
      "Sequencing Saturation,0.250247\n",
      "Q30 Bases in CB+UMI,0.795757\n",
      "Q30 Bases in RNA read,0.757655\n",
      "Reads Mapped to Genome: Unique+Multiple,0.492646\n",
      "Reads Mapped to Genome: Unique,0.492646\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.297433\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.289762\n",
      "Estimated Number of Cells,13386\n",
      "Reads in Cells Mapped to Unique Genes,20666879\n",
      "Fraction of Reads in Cells,0.593506\n",
      "Mean Reads per Cell,1543\n",
      "Median Reads per Cell,1289\n",
      "UMIs in Cells,15190497\n",
      "Mean UMI per Cell,1134\n",
      "Median UMI per Cell,963\n",
      "Mean Genes per Cell,842\n",
      "Median Genes per Cell,755\n",
      "Total Genes Detected,20932\n",
      "\n",
      "SRR10545078.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,121251623\n",
      "Reads With Valid Barcodes,0.411885\n",
      "Sequencing Saturation,0.24297\n",
      "Q30 Bases in CB+UMI,0.785203\n",
      "Q30 Bases in RNA read,0.752037\n",
      "Reads Mapped to Genome: Unique+Multiple,0.490856\n",
      "Reads Mapped to Genome: Unique,0.490856\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.28258\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.275263\n",
      "Estimated Number of Cells,13247\n",
      "Reads in Cells Mapped to Unique Genes,19696659\n",
      "Fraction of Reads in Cells,0.590144\n",
      "Mean Reads per Cell,1486\n",
      "Median Reads per Cell,1241\n",
      "UMIs in Cells,14612934\n",
      "Mean UMI per Cell,1103\n",
      "Median UMI per Cell,936\n",
      "Mean Genes per Cell,822\n",
      "Median Genes per Cell,735\n",
      "Total Genes Detected,20880\n",
      "\n",
      "SRR10545079.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,100751546\n",
      "Reads With Valid Barcodes,0.407114\n",
      "Sequencing Saturation,0.200409\n",
      "Q30 Bases in CB+UMI,0.737365\n",
      "Q30 Bases in RNA read,0.707794\n",
      "Reads Mapped to Genome: Unique+Multiple,0.454444\n",
      "Reads Mapped to Genome: Unique,0.454444\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.265971\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.258983\n",
      "Estimated Number of Cells,12834\n",
      "Reads in Cells Mapped to Unique Genes,14979983\n",
      "Fraction of Reads in Cells,0.574102\n",
      "Mean Reads per Cell,1167\n",
      "Median Reads per Cell,971\n",
      "UMIs in Cells,11792879\n",
      "Mean UMI per Cell,918\n",
      "Median UMI per Cell,774\n",
      "Mean Genes per Cell,706\n",
      "Median Genes per Cell,625\n",
      "Total Genes Detected,20515\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for sample in $dir/*/Gene/Summary.csv\n",
    "do\n",
    "    echo ${sample#*/}\n",
    "    cat $sample\n",
    "    cut -d, -f2 $sample > ${sample%/Gene/Summary.csv}/Gene/Summary_numbers.csv\n",
    "    printf \"\\n\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
