{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Most of the steps below have been copied from the scanpy tutorial: https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.8.1 anndata==0.7.5 umap==0.5.1 numpy==1.20.2 scipy==1.6.2 pandas==1.3.3 scikit-learn==0.24.1 statsmodels==0.12.2 python-igraph==0.9.0 leidenalg==0.8.3 pynndescent==0.5.2\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=150, facecolor=\"white\", vector_friendly=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make sure that the packages that you have match the ones below IF you want to verify our results and obtain exactly the same results as described below. If you want to simply analyse your own dataset, then most likely later versions will also be fine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Package                            Version\n",
      "---------------------------------- -----------------\n",
      "adjustText                         0.7.3\n",
      "aiohttp                            3.7.4.post0\n",
      "aiohttp-cors                       0.7.0\n",
      "aioredis                           1.3.1\n",
      "alabaster                          0.7.12\n",
      "anaconda-client                    1.7.2\n",
      "anaconda-navigator                 1.9.7\n",
      "anaconda-project                   0.8.3\n",
      "anndata                            0.7.5\n",
      "annoy                              1.17.0\n",
      "ansiwrap                           0.8.4\n",
      "anyio                              2.2.0\n",
      "appdirs                            1.4.4\n",
      "arboreto                           0.1.6\n",
      "argcomplete                        1.12.2\n",
      "argh                               0.26.2\n",
      "argon2-cffi                        20.1.0\n",
      "asciitree                          0.3.3\n",
      "asn1crypto                         1.4.0\n",
      "astroid                            2.5\n",
      "astropy                            3.2.1\n",
      "async-timeout                      3.0.1\n",
      "atomicwrites                       1.4.0\n",
      "attrs                              20.3.0\n",
      "Babel                              2.9.0\n",
      "backcall                           0.2.0\n",
      "backports.functools-lru-cache      1.6.1\n",
      "backports.os                       0.1.1\n",
      "backports.shutil-get-terminal-size 1.0.0\n",
      "backports.tempfile                 1.0\n",
      "backports.weakref                  1.0.post1\n",
      "bap-atac                           0.6.7\n",
      "bash-kernel                        0.7.2\n",
      "bbknn                              1.4.0\n",
      "beautifulsoup4                     4.9.3\n",
      "bgzip                              0.3.5\n",
      "biopython                          1.77\n",
      "bitarray                           1.6.3\n",
      "bkcharts                           0.2\n",
      "black                              20.8b1\n",
      "bleach                             3.3.0\n",
      "blessings                          1.7\n",
      "blis                               0.7.4\n",
      "bokeh                              2.3.0\n",
      "boltons                            20.2.1\n",
      "boto                               2.49.0\n",
      "Bottleneck                         1.3.2\n",
      "brotlipy                           0.7.0\n",
      "cached-property                    1.5.2\n",
      "cachetools                         4.2.1\n",
      "catalogue                          1.0.0\n",
      "certifi                            2021.10.8\n",
      "cffi                               1.14.5\n",
      "chardet                            4.0.0\n",
      "click                              7.1.2\n",
      "cloudpickle                        1.6.0\n",
      "clyent                             1.2.2\n",
      "colorama                           0.4.4\n",
      "colorful                           0.5.4\n",
      "conda                              4.11.0\n",
      "conda-build                        3.18.8\n",
      "conda-package-handling             1.7.2\n",
      "conda-verify                       3.4.2\n",
      "ConfigArgParse                     1.2.3\n",
      "contextlib2                        0.6.0.post1\n",
      "cooler                             0.8.11\n",
      "cryptography                       3.3.1\n",
      "ctxcore                            0.1.1\n",
      "cycler                             0.10.0\n",
      "cymem                              2.0.5\n",
      "Cython                             0.29.22\n",
      "cytoolz                            0.10.1\n",
      "dask                               2021.3.1\n",
      "datrie                             0.8.2\n",
      "decorator                          4.4.2\n",
      "deepTools                          3.5.1\n",
      "deeptoolsintervals                 0.1.9\n",
      "defusedxml                         0.6.0\n",
      "deprecation                        2.1.0\n",
      "dill                               0.3.4.dev0\n",
      "distributed                        2021.3.1\n",
      "docutils                           0.16\n",
      "entrypoints                        0.3\n",
      "et-xmlfile                         1.0.1\n",
      "fastcache                          1.1.0\n",
      "fastkde                            1.0.19\n",
      "fbpca                              1.0\n",
      "filelock                           3.0.12\n",
      "Flask                              1.1.2\n",
      "frozendict                         1.2\n",
      "fsspec                             0.8.7\n",
      "future                             0.18.2\n",
      "fuzzysearch                        0.7.3\n",
      "gensim                             3.8.3\n",
      "geosketch                          1.2\n",
      "germalemma                         0.1.3\n",
      "get-version                        2.1\n",
      "gevent                             21.1.1\n",
      "gitdb                              4.0.5\n",
      "GitPython                          3.1.3\n",
      "glob2                              0.7\n",
      "globre                             0.1.5\n",
      "gmpy2                              2.0.8\n",
      "google-api-core                    1.26.1\n",
      "google-auth                        1.27.1\n",
      "googleapis-common-protos           1.53.0\n",
      "gpustat                            0.6.0\n",
      "greenlet                           1.0.0\n",
      "grpcio                             1.36.1\n",
      "gspread                            3.1.0\n",
      "h5py                               3.2.1\n",
      "harmonypy                          0.0.5\n",
      "HeapDict                           1.0.1\n",
      "hiredis                            1.1.0\n",
      "html5lib                           1.1\n",
      "httplib2                           0.17.0\n",
      "idna                               2.10\n",
      "imageio                            2.9.0\n",
      "imagesize                          1.2.0\n",
      "importlib-metadata                 3.10.0\n",
      "intel-openmp                       2020.0.133\n",
      "interlap                           0.2.7\n",
      "intervaltree                       2.1.0\n",
      "ipykernel                          5.5.0\n",
      "ipympl                             0.7.0\n",
      "ipython                            7.21.0\n",
      "ipython-genutils                   0.2.0\n",
      "ipywidgets                         7.6.3\n",
      "isort                              5.7.0\n",
      "itsdangerous                       1.1.0\n",
      "jdcal                              1.4.1\n",
      "jedi                               0.18.0\n",
      "jeepney                            0.6.0\n",
      "jgraph                             0.2.1\n",
      "Jinja2                             2.11.3\n",
      "joblib                             1.0.1\n",
      "json5                              0.9.5\n",
      "jsonschema                         3.0.1\n",
      "jupyter                            1.0.0\n",
      "jupyter-client                     6.1.12\n",
      "jupyter-console                    6.2.0\n",
      "jupyter-contrib-core               0.3.3\n",
      "jupyter-contrib-nbextensions       0.5.1\n",
      "jupyter-core                       4.7.1\n",
      "jupyter-highlight-selected-word    0.2.0\n",
      "jupyter-latex-envs                 1.4.6\n",
      "jupyter-nbextensions-configurator  0.4.1\n",
      "jupyter-packaging                  0.7.12\n",
      "jupyter-server                     1.5.1\n",
      "jupyterlab                         3.0.12\n",
      "jupyterlab-code-formatter          1.4.5\n",
      "jupyterlab-server                  2.3.0\n",
      "jupyterlab-widgets                 1.0.0\n",
      "keyring                            22.2.0\n",
      "kiwisolver                         1.3.1\n",
      "lazy-object-proxy                  1.5.2\n",
      "lda                                2.0.0\n",
      "legacy-api-wrap                    1.2\n",
      "leidenalg                          0.8.3\n",
      "libarchive-c                       2.9\n",
      "llvmlite                           0.36.0\n",
      "locket                             0.2.1\n",
      "loompy                             3.0.6\n",
      "loomxpy                            0.1.0\n",
      "lxml                               4.6.2\n",
      "MACS2                              2.2.6\n",
      "MarkupSafe                         1.1.1\n",
      "matplotlib                         3.3.4\n",
      "mccabe                             0.6.1\n",
      "mistune                            0.8.4\n",
      "mkl                                2019.0\n",
      "mkl-fft                            1.3.0\n",
      "mkl-random                         1.1.1\n",
      "mkl-service                        2.3.0\n",
      "mock                               4.0.3\n",
      "more-itertools                     8.7.0\n",
      "mpmath                             1.1.0\n",
      "msgpack                            1.0.2\n",
      "multidict                          5.1.0\n",
      "multipledispatch                   0.6.0\n",
      "multiprocess                       0.70.11.1\n",
      "multiprocessing-on-dill            3.5.0a4\n",
      "murmurhash                         1.0.5\n",
      "mypy-extensions                    0.4.3\n",
      "natsort                            7.1.1\n",
      "navigator-updater                  0.2.1\n",
      "nbclassic                          0.2.6\n",
      "nbclient                           0.5.4\n",
      "nbconvert                          5.5.0\n",
      "nbformat                           5.1.3\n",
      "ncls                               0.0.57\n",
      "nest-asyncio                       1.5.1\n",
      "networkx                           2.5\n",
      "nltk                               3.5\n",
      "nose                               1.3.7\n",
      "notebook                           6.0.0\n",
      "numba                              0.53.1\n",
      "numexpr                            2.7.2\n",
      "numpy                              1.20.2\n",
      "numpy-groupies                     0.9.13\n",
      "numpydoc                           1.1.0\n",
      "nvidia-ml-py3                      7.352.0\n",
      "oauth2client                       4.1.3\n",
      "olefile                            0.46\n",
      "opencensus                         0.7.12\n",
      "opencensus-context                 0.1.2\n",
      "openpyxl                           3.0.6\n",
      "optparse-pretty                    0.1.1\n",
      "packaging                          20.9\n",
      "pandas                             1.3.3\n",
      "pandocfilters                      1.4.3\n",
      "papermill                          2.3.3\n",
      "parso                              0.8.1\n",
      "partd                              1.1.0\n",
      "path                               15.1.2\n",
      "pathlib2                           2.3.5\n",
      "pathos                             0.2.8.dev0\n",
      "pathspec                           0.8.1\n",
      "patsy                              0.5.1\n",
      "PatternLite                        3.6\n",
      "pbr                                3.1.1\n",
      "pep8                               1.7.1\n",
      "pexpect                            4.8.0\n",
      "pickleshare                        0.7.5\n",
      "Pillow                             8.1.2\n",
      "pip                                21.0.1\n",
      "pkginfo                            1.7.0\n",
      "plac                               1.1.3\n",
      "plotly                             4.14.3\n",
      "pluggy                             0.13.1\n",
      "ply                                3.11\n",
      "pox                                0.2.9\n",
      "ppft                               1.6.6.3\n",
      "preshed                            3.0.5\n",
      "prometheus-client                  0.9.0\n",
      "prompt-toolkit                     3.0.17\n",
      "protobuf                           3.15.5\n",
      "psutil                             5.8.0\n",
      "ptyprocess                         0.7.0\n",
      "py                                 1.10.0\n",
      "py-spy                             0.3.4\n",
      "py2bit                             0.3.0\n",
      "pyarrow                            0.16.0\n",
      "pyasn1                             0.4.8\n",
      "pyasn1-modules                     0.2.8\n",
      "pybedtools                         0.8.1\n",
      "pyBigWig                           0.3.18\n",
      "pybiomart                          0.2.0\n",
      "pycodestyle                        2.6.0\n",
      "pycosat                            0.6.3\n",
      "pycparser                          2.20\n",
      "pycrypto                           2.6.1\n",
      "pycurl                             7.43.0.6\n",
      "pyfaidx                            0.5.9.5\n",
      "pyfasta                            0.5.2\n",
      "pyflakes                           2.2.0\n",
      "Pygments                           2.8.0\n",
      "pylint                             2.7.1\n",
      "pynndescent                        0.5.2\n",
      "pyodbc                             4.0.0-unsupported\n",
      "pyOpenSSL                          20.0.1\n",
      "pypairix                           0.3.7\n",
      "pyparsing                          2.4.7\n",
      "Pyphen                             0.10.0\n",
      "pyranges                           0.0.95\n",
      "pyrle                              0.0.32\n",
      "pyrsistent                         0.17.3\n",
      "pysam                              0.15.3\n",
      "pyscenic                           0.11.2\n",
      "PySocks                            1.7.1\n",
      "pytest                             5.0.1\n",
      "pytest-arraydiff                   0.3\n",
      "pytest-astropy                     0.5.0\n",
      "pytest-doctestplus                 0.3.0\n",
      "pytest-openfiles                   0.3.2\n",
      "pytest-remotedata                  0.3.1\n",
      "python-dateutil                    2.8.1\n",
      "python-igraph                      0.9.0\n",
      "python-Levenshtein                 0.12.2\n",
      "python-louvain                     0.15\n",
      "python-oauth2                      1.1.1\n",
      "pytz                               2021.1\n",
      "PyWavelets                         1.1.1\n",
      "PyYAML                             5.4.1\n",
      "pyzmq                              22.0.3\n",
      "QtAwesome                          1.0.1\n",
      "qtconsole                          5.0.2\n",
      "QtPy                               1.9.0\n",
      "ratelimiter                        1.2.0.post0\n",
      "ray                                1.2.0\n",
      "redis                              3.5.3\n",
      "regex                              2020.11.13\n",
      "requests                           2.25.1\n",
      "requests-cache                     0.5.2\n",
      "retrying                           1.3.3\n",
      "rope                               0.18.0\n",
      "rsa                                4.7.2\n",
      "ruamel-yaml                        0.15.87\n",
      "ruamel.yaml                        0.16.10\n",
      "ruamel.yaml.clib                   0.2.0\n",
      "scanorama                          1.7.1\n",
      "scanpy                             1.8.1\n",
      "scikit-image                       0.17.2\n",
      "scikit-learn                       0.24.1\n",
      "scipy                              1.6.2\n",
      "scrublet                           0.2.3\n",
      "seaborn                            0.11.1\n",
      "SecretStorage                      3.3.1\n",
      "Send2Trash                         1.5.0\n",
      "setuptools                         54.2.0\n",
      "simplegeneric                      0.8.1\n",
      "simplejson                         3.17.2\n",
      "sinfo                              0.3.1\n",
      "singledispatch                     0.0.0\n",
      "sinto                              0.7.1\n",
      "six                                1.15.0\n",
      "sklearn                            0.0\n",
      "smart-open                         4.2.0\n",
      "smmap                              3.0.4\n",
      "snakemake                          5.19.3\n",
      "snaptools                          1.4.8\n",
      "sniffio                            1.2.0\n",
      "snowballstemmer                    2.1.0\n",
      "sorted-nearest                     0.0.32\n",
      "sortedcollections                  2.1.0\n",
      "sortedcontainers                   2.3.0\n",
      "soupsieve                          2.2\n",
      "spacy                              2.3.5\n",
      "Sphinx                             3.5.1\n",
      "sphinxcontrib-applehelp            1.0.2\n",
      "sphinxcontrib-devhelp              1.0.2\n",
      "sphinxcontrib-htmlhelp             1.0.3\n",
      "sphinxcontrib-jsmath               1.0.1\n",
      "sphinxcontrib-qthelp               1.0.3\n",
      "sphinxcontrib-serializinghtml      1.1.4\n",
      "sphinxcontrib-websupport           1.2.4\n",
      "spyder                             3.3.6\n",
      "spyder-kernels                     0.5.1\n",
      "SQLAlchemy                         1.3.23\n",
      "srsly                              1.0.5\n",
      "statsmodels                        0.12.2\n",
      "stdlib-list                        0.8.0\n",
      "subsample                          0.0.6\n",
      "sympy                              1.7.1\n",
      "tables                             3.6.1\n",
      "tabulate                           0.8.9\n",
      "tblib                              1.7.0\n",
      "tenacity                           8.0.1\n",
      "terminado                          0.9.2\n",
      "testpath                           0.4.4\n",
      "texttable                          1.6.3\n",
      "textwrap3                          0.9.2\n",
      "thinc                              7.4.5\n",
      "threadpoolctl                      2.1.0\n",
      "tifffile                           2020.10.1\n",
      "tmtoolkit                          0.9.0\n",
      "toml                               0.10.1\n",
      "toolz                              0.11.1\n",
      "toposort                           1.5\n",
      "tornado                            6.1\n",
      "tqdm                               4.59.0\n",
      "traitlets                          5.0.5\n",
      "typed-ast                          1.4.2\n",
      "typing                             3.7.4.3\n",
      "typing-extensions                  3.7.4.3\n",
      "umap                               0.1.1\n",
      "umap-learn                         0.5.1\n",
      "uncertainties                      3.1.2\n",
      "unicodecsv                         0.14.1\n",
      "urllib3                            1.26.4\n",
      "wasabi                             0.8.2\n",
      "wcwidth                            0.2.5\n",
      "webencodings                       0.5.1\n",
      "Werkzeug                           1.0.1\n",
      "wheel                              0.36.2\n",
      "widgetsnbextension                 3.5.1\n",
      "wrapt                              1.12.1\n",
      "wurlitzer                          2.0.1\n",
      "xlrd                               2.0.1\n",
      "XlsxWriter                         1.3.7\n",
      "xlwt                               1.3.0\n",
      "yarl                               1.6.3\n",
      "zict                               2.0.0\n",
      "zipp                               3.4.1\n",
      "zope.event                         4.5.0\n",
      "zope.interface                     5.2.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# drop-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['demultiplexed_dropseq_sub/SRR1853178.Solo.out/Gene/raw/',\n",
       " 'demultiplexed_dropseq_sub/SRR1853179.Solo.out/Gene/raw/',\n",
       " 'demultiplexed_dropseq_sub/SRR1853180.Solo.out/Gene/raw/',\n",
       " 'demultiplexed_dropseq_sub/SRR1853181.Solo.out/Gene/raw/',\n",
       " 'demultiplexed_dropseq_sub/SRR1853182.Solo.out/Gene/raw/',\n",
       " 'demultiplexed_dropseq_sub/SRR1853183.Solo.out/Gene/raw/',\n",
       " 'demultiplexed_dropseq_sub/SRR1853184.Solo.out/Gene/raw/']"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx_dirs = sorted(glob.glob('demultiplexed_dropseq_sub/*Solo.out/Gene/raw/'))\n",
    "mtx_dirs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncells_dict = {\n",
    "    'SRR1853178': 6600,\n",
    "    'SRR1853179': 9000,\n",
    "    'SRR1853180': 6120,\n",
    "    'SRR1853181': 7650,\n",
    "    'SRR1853182': 7650,\n",
    "    'SRR1853183': 8280,\n",
    "    'SRR1853184': 4000,\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for mtx_dir in mtx_dirs:\n",
    "    sample=mtx_dir.replace('demultiplexed_dropseq_sub/', '')\n",
    "    sample=sample.replace('.Solo.out/Gene/raw/', '')\n",
    "    adata = sc.read_10x_mtx(\n",
    "        mtx_dir,  # the directory with the `.mtx` file\n",
    "        var_names=\"gene_symbols\",  # use gene symbols for the variable names (variables-axis index)\n",
    "        cache=True,\n",
    "        )  # write a cache file for faster subsequent reading\n",
    "    adata.obs_names = sample + '_' + adata.obs_names\n",
    "    adata.var_names_make_unique()\n",
    "    sc.pp.calculate_qc_metrics(adata, inplace=True)\n",
    "    adata.obs.sort_values(by='n_genes_by_counts', ascending=False, inplace=True)\n",
    "    ncells = ncells_dict[sample]\n",
    "    adata = adata[0:ncells-1]\n",
    "    adata.obs.to_csv(sample + '_adata.tsv', sep='\\t', header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge into one nice dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged = pd.DataFrame()\n",
    "for file in glob.glob('tables/dropseq_SRR18531*_adata.tsv'):\n",
    "    df = pd.read_csv(file, sep='\\t', index_col=0)\n",
    "    df_merged = df_merged.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR1853184_CTTTACTTGGTC</th>\n",
       "      <td>7713</td>\n",
       "      <td>8.950792</td>\n",
       "      <td>38554.0</td>\n",
       "      <td>10.559841</td>\n",
       "      <td>14.229393</td>\n",
       "      <td>19.248327</td>\n",
       "      <td>26.280023</td>\n",
       "      <td>39.741661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1853184_CCGCTCTAAAAG</th>\n",
       "      <td>7506</td>\n",
       "      <td>8.923591</td>\n",
       "      <td>45529.0</td>\n",
       "      <td>10.726127</td>\n",
       "      <td>21.280942</td>\n",
       "      <td>27.332030</td>\n",
       "      <td>34.402249</td>\n",
       "      <td>46.945903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1853184_CGGCACAGTGGA</th>\n",
       "      <td>7316</td>\n",
       "      <td>8.897956</td>\n",
       "      <td>33693.0</td>\n",
       "      <td>10.425076</td>\n",
       "      <td>14.982341</td>\n",
       "      <td>20.223785</td>\n",
       "      <td>27.658564</td>\n",
       "      <td>41.041166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1853184_GATGCCGGAATG</th>\n",
       "      <td>6887</td>\n",
       "      <td>8.837536</td>\n",
       "      <td>23715.0</td>\n",
       "      <td>10.073905</td>\n",
       "      <td>10.221379</td>\n",
       "      <td>15.306768</td>\n",
       "      <td>22.230656</td>\n",
       "      <td>35.686275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1853184_AGGGCTCGCGAC</th>\n",
       "      <td>6351</td>\n",
       "      <td>8.756525</td>\n",
       "      <td>29061.0</td>\n",
       "      <td>10.277186</td>\n",
       "      <td>20.395031</td>\n",
       "      <td>26.520078</td>\n",
       "      <td>34.011218</td>\n",
       "      <td>47.338357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1853181_GTCCGCGTCTCT</th>\n",
       "      <td>383</td>\n",
       "      <td>5.950643</td>\n",
       "      <td>547.0</td>\n",
       "      <td>6.306275</td>\n",
       "      <td>33.638026</td>\n",
       "      <td>48.263254</td>\n",
       "      <td>66.544790</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1853181_GGAAAACGACTG</th>\n",
       "      <td>383</td>\n",
       "      <td>5.950643</td>\n",
       "      <td>574.0</td>\n",
       "      <td>6.354370</td>\n",
       "      <td>33.972125</td>\n",
       "      <td>50.696864</td>\n",
       "      <td>68.118467</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1853181_GCCCAGAAATCT</th>\n",
       "      <td>383</td>\n",
       "      <td>5.950643</td>\n",
       "      <td>454.0</td>\n",
       "      <td>6.120297</td>\n",
       "      <td>26.651982</td>\n",
       "      <td>37.665198</td>\n",
       "      <td>59.691630</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1853181_TGGTGGGATCAC</th>\n",
       "      <td>383</td>\n",
       "      <td>5.950643</td>\n",
       "      <td>520.0</td>\n",
       "      <td>6.255750</td>\n",
       "      <td>31.730769</td>\n",
       "      <td>45.576923</td>\n",
       "      <td>64.807692</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1853181_GGCACAGAGCCA</th>\n",
       "      <td>383</td>\n",
       "      <td>5.950643</td>\n",
       "      <td>477.0</td>\n",
       "      <td>6.169611</td>\n",
       "      <td>30.188679</td>\n",
       "      <td>40.670860</td>\n",
       "      <td>61.635220</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>49293 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         n_genes_by_counts  log1p_n_genes_by_counts  \\\n",
       "SRR1853184_CTTTACTTGGTC               7713                 8.950792   \n",
       "SRR1853184_CCGCTCTAAAAG               7506                 8.923591   \n",
       "SRR1853184_CGGCACAGTGGA               7316                 8.897956   \n",
       "SRR1853184_GATGCCGGAATG               6887                 8.837536   \n",
       "SRR1853184_AGGGCTCGCGAC               6351                 8.756525   \n",
       "...                                    ...                      ...   \n",
       "SRR1853181_GTCCGCGTCTCT                383                 5.950643   \n",
       "SRR1853181_GGAAAACGACTG                383                 5.950643   \n",
       "SRR1853181_GCCCAGAAATCT                383                 5.950643   \n",
       "SRR1853181_TGGTGGGATCAC                383                 5.950643   \n",
       "SRR1853181_GGCACAGAGCCA                383                 5.950643   \n",
       "\n",
       "                         total_counts  log1p_total_counts  \\\n",
       "SRR1853184_CTTTACTTGGTC       38554.0           10.559841   \n",
       "SRR1853184_CCGCTCTAAAAG       45529.0           10.726127   \n",
       "SRR1853184_CGGCACAGTGGA       33693.0           10.425076   \n",
       "SRR1853184_GATGCCGGAATG       23715.0           10.073905   \n",
       "SRR1853184_AGGGCTCGCGAC       29061.0           10.277186   \n",
       "...                               ...                 ...   \n",
       "SRR1853181_GTCCGCGTCTCT         547.0            6.306275   \n",
       "SRR1853181_GGAAAACGACTG         574.0            6.354370   \n",
       "SRR1853181_GCCCAGAAATCT         454.0            6.120297   \n",
       "SRR1853181_TGGTGGGATCAC         520.0            6.255750   \n",
       "SRR1853181_GGCACAGAGCCA         477.0            6.169611   \n",
       "\n",
       "                         pct_counts_in_top_50_genes  \\\n",
       "SRR1853184_CTTTACTTGGTC                   14.229393   \n",
       "SRR1853184_CCGCTCTAAAAG                   21.280942   \n",
       "SRR1853184_CGGCACAGTGGA                   14.982341   \n",
       "SRR1853184_GATGCCGGAATG                   10.221379   \n",
       "SRR1853184_AGGGCTCGCGAC                   20.395031   \n",
       "...                                             ...   \n",
       "SRR1853181_GTCCGCGTCTCT                   33.638026   \n",
       "SRR1853181_GGAAAACGACTG                   33.972125   \n",
       "SRR1853181_GCCCAGAAATCT                   26.651982   \n",
       "SRR1853181_TGGTGGGATCAC                   31.730769   \n",
       "SRR1853181_GGCACAGAGCCA                   30.188679   \n",
       "\n",
       "                         pct_counts_in_top_100_genes  \\\n",
       "SRR1853184_CTTTACTTGGTC                    19.248327   \n",
       "SRR1853184_CCGCTCTAAAAG                    27.332030   \n",
       "SRR1853184_CGGCACAGTGGA                    20.223785   \n",
       "SRR1853184_GATGCCGGAATG                    15.306768   \n",
       "SRR1853184_AGGGCTCGCGAC                    26.520078   \n",
       "...                                              ...   \n",
       "SRR1853181_GTCCGCGTCTCT                    48.263254   \n",
       "SRR1853181_GGAAAACGACTG                    50.696864   \n",
       "SRR1853181_GCCCAGAAATCT                    37.665198   \n",
       "SRR1853181_TGGTGGGATCAC                    45.576923   \n",
       "SRR1853181_GGCACAGAGCCA                    40.670860   \n",
       "\n",
       "                         pct_counts_in_top_200_genes  \\\n",
       "SRR1853184_CTTTACTTGGTC                    26.280023   \n",
       "SRR1853184_CCGCTCTAAAAG                    34.402249   \n",
       "SRR1853184_CGGCACAGTGGA                    27.658564   \n",
       "SRR1853184_GATGCCGGAATG                    22.230656   \n",
       "SRR1853184_AGGGCTCGCGAC                    34.011218   \n",
       "...                                              ...   \n",
       "SRR1853181_GTCCGCGTCTCT                    66.544790   \n",
       "SRR1853181_GGAAAACGACTG                    68.118467   \n",
       "SRR1853181_GCCCAGAAATCT                    59.691630   \n",
       "SRR1853181_TGGTGGGATCAC                    64.807692   \n",
       "SRR1853181_GGCACAGAGCCA                    61.635220   \n",
       "\n",
       "                         pct_counts_in_top_500_genes  \n",
       "SRR1853184_CTTTACTTGGTC                    39.741661  \n",
       "SRR1853184_CCGCTCTAAAAG                    46.945903  \n",
       "SRR1853184_CGGCACAGTGGA                    41.041166  \n",
       "SRR1853184_GATGCCGGAATG                    35.686275  \n",
       "SRR1853184_AGGGCTCGCGAC                    47.338357  \n",
       "...                                              ...  \n",
       "SRR1853181_GTCCGCGTCTCT                   100.000000  \n",
       "SRR1853181_GGAAAACGACTG                   100.000000  \n",
       "SRR1853181_GCCCAGAAATCT                   100.000000  \n",
       "SRR1853181_TGGTGGGATCAC                   100.000000  \n",
       "SRR1853181_GGCACAGAGCCA                   100.000000  \n",
       "\n",
       "[49293 rows x 8 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "n_genes_by_counts              671.000000\n",
       "log1p_n_genes_by_counts          6.510258\n",
       "total_counts                   952.000000\n",
       "log1p_total_counts               6.859615\n",
       "pct_counts_in_top_50_genes      23.401950\n",
       "pct_counts_in_top_100_genes     34.064213\n",
       "pct_counts_in_top_200_genes     49.536178\n",
       "pct_counts_in_top_500_genes     81.986900\n",
       "dtype: float64"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged.median()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I can now also write a drop-seq barcode list!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged['barcode'] = [df_merged.index[x].split('_')[1] for x in range(len(df_merged.index))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged['barcode'].to_csv('drop-seq_macosko_whitelist.txt', index=False, header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using uniq -u $whitelist | wc -l, I found that this barcode list is unique. Remap using this whitelist now."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# indrop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For indrop, we take a slightly different approach. We mapped all the pooled data. We read the large unfiltered matrix. We then take the top 1 428 898 458 reads / 52738 = 27094 cells to match hydrop's read depth."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx_dir = 'demultiplexed_indrop_index/merged.Solo.out/Gene/raw/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... reading from cache file cache/demultiplexed_indrop_index-merged.Solo.out-Gene-raw-matrix.h5ad\n"
     ]
    }
   ],
   "source": [
    "sample = 'merged'\n",
    "adata = sc.read_10x_mtx(\n",
    "    mtx_dir,  # the directory with the `.mtx` file\n",
    "    var_names=\"gene_symbols\",  # use gene symbols for the variable names (variables-axis index)\n",
    "    cache=True,\n",
    "    )  # write a cache file for faster subsequent reading\n",
    "adata.obs_names = sample + '_' + adata.obs_names\n",
    "adata.var_names_make_unique()\n",
    "sc.pp.calculate_qc_metrics(adata, inplace=True)\n",
    "adata.obs.sort_values(by='n_genes_by_counts', ascending=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[0:27094-1]\n",
    "adata.obs.to_csv('indrop' + sample + '_adata.tsv', sep='\\t', header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "n_genes_by_counts              1321.000000\n",
       "log1p_n_genes_by_counts           7.186901\n",
       "total_counts                   1920.000000\n",
       "log1p_total_counts                7.560601\n",
       "pct_counts_in_top_50_genes       17.385621\n",
       "pct_counts_in_top_100_genes      25.818918\n",
       "pct_counts_in_top_200_genes      37.254902\n",
       "pct_counts_in_top_500_genes      58.153242\n",
       "dtype: float64"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
