{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aa6ca910",
   "metadata": {},
   "source": [
    "We take as a whitelist the cell barcodes accepted in the first mapping."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eb50adf0",
   "metadata": {
    "papermill": {
     "duration": 1.85807,
     "end_time": "2022-01-31T00:30:47.685974",
     "exception": false,
     "start_time": "2022-01-31T00:30:45.827904",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "star_solo_hydrop_rna_mapping_unique_cellfilter_dropseq () {\n",
    "    local fastq_R1_filename=\"${1}\";\n",
    "    local fastq_R2_filename=\"${2}\";\n",
    "    local bam_filename=\"${3}\";\n",
    "    star_reference_dir='/lustre1/project/stg_00002/lcb/fderop/data/00000000_genomes/mm10_STAR_2.7.8_premrna';\n",
    "    sjdbgtf='/lustre1/project/stg_00002/lcb/fderop/data/00000000_genomes/mm10_STAR_2.7.8_premrna/genes.gtf'\n",
    "    #module load STAR/2.7.5b-foss-2018a;\n",
    "    #module load SAMtools/1.10-foss-2018a;\n",
    "    # Map with STAR solo\n",
    "    /lustre1/project/stg_00002/lcb/fderop/scripts/STAR-2.7.8a/bin/Linux_x86_64/STAR \\\n",
    "        --runThreadN 24 \\\n",
    "        --runMode alignReads \\\n",
    "        --outSAMtype BAM SortedByCoordinate \\\n",
    "        --sysShell /bin/bash \\\n",
    "        --genomeDir \"${star_reference_dir}\" \\\n",
    "        --readFilesIn \"${fastq_R1_filename}\" \"${fastq_R2_filename}\" \\\n",
    "        --readFilesCommand 'gzip -c -d' \\\n",
    "        --soloCBwhitelist '/lustre1/project/stg_00002/lcb/fderop/data/20220125_hydrop_data_analysis/HyDrop-RNA_publicdata_comparison/dropseq_whitelist/drop-seq_macosko_whitelist.txt' \\\n",
    "        --soloType CB_UMI_Simple \\\n",
    "        --soloCBmatchWLtype 1MM \\\n",
    "        --soloCBstart  1 \\\n",
    "        --soloCBlen 12 \\\n",
    "        --soloUMIstart 13 \\\n",
    "        --soloUMIlen 8\\\n",
    "        --sjdbGTFfile $sjdbgtf \\\n",
    "        --soloCellFilter CellRanger2.2 2000 0.99 10 \\\n",
    "        --outFilterMultimapNmax 1 \\\n",
    "        --outSAMattributes NH HI AS nM CB UB CR CY UR UY \\\n",
    "        --outFileNamePrefix ${bam_filename%bam} \\\n",
    "        --outReadsUnmapped Fastx \\\n",
    "        --quantMode GeneCounts \\\n",
    "        --bamRemoveDuplicatesType UniqueIdentical \\\n",
    "        --soloFeatures Gene GeneFull\n",
    "        # Index BAM file.\n",
    "    # samtools index \"${bam_filename%bam}Aligned.sortedByCoord.out.bam\"\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "800b9bd5",
   "metadata": {
    "papermill": {
     "duration": 0.005748,
     "end_time": "2022-01-31T00:30:47.729894",
     "exception": false,
     "start_time": "2022-01-31T00:30:47.724146",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Now we run this star solo script in a for loop on all the R2 fastq files in `fastq/`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a44484c4",
   "metadata": {
    "papermill": {
     "duration": 18900.363004,
     "end_time": "2022-01-31T05:45:48.097684",
     "exception": false,
     "start_time": "2022-01-31T00:30:47.734680",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fastq_dropseq_sub/SRR1853178\n",
      "Feb 01 13:05:16 ..... started STAR run\n",
      "Feb 01 13:05:16 ..... loading genome\n",
      "Feb 01 13:05:50 ..... processing annotations GTF\n",
      "Feb 01 13:05:56 ..... started mapping\n",
      "Feb 01 13:33:33 ..... finished mapping\n",
      "Feb 01 13:33:37 ..... started Solo counting\n",
      "Feb 01 13:38:09 ..... finished Solo counting\n",
      "Feb 01 13:38:09 ..... started sorting BAM\n",
      "Feb 01 13:44:36 ..... finished successfully\n",
      "fastq_dropseq_sub/SRR1853179\n",
      "Feb 01 13:44:53 ..... started STAR run\n",
      "Feb 01 13:44:54 ..... loading genome\n",
      "Feb 01 13:45:26 ..... processing annotations GTF\n",
      "Feb 01 13:45:31 ..... started mapping\n",
      "Feb 01 14:24:20 ..... finished mapping\n",
      "Feb 01 14:24:24 ..... started Solo counting\n",
      "Feb 01 14:32:55 ..... finished Solo counting\n",
      "Feb 01 14:32:55 ..... started sorting BAM\n"
     ]
    }
   ],
   "source": [
    "dir=demultiplexed_dropseq_sub_whitelist\n",
    "mkdir $dir\n",
    "for sample in fastq_dropseq_sub/*_2.SUB.fastq.gz\n",
    "do\n",
    "    NAME=${sample%_2.SUB.fastq.gz}\n",
    "    OUT=${NAME##*/}\n",
    "    echo $NAME\n",
    "    star_solo_hydrop_rna_mapping_unique_cellfilter_dropseq \\\n",
    "        ${NAME}_2.SUB.fastq.gz \\\n",
    "        ${NAME}_1.SUB.fastq.gz \\\n",
    "        $dir/$OUT.bam\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b3707ef2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘demultiplexed_dropseq_sub_whitelist’: File exists\n",
      "fastq_dropseq_sub/SRR1853182\n",
      "Feb 01 16:54:40 ..... started STAR run\n",
      "Feb 01 16:54:42 ..... loading genome\n",
      "Feb 01 16:55:34 ..... processing annotations GTF\n",
      "Feb 01 16:55:40 ..... started mapping\n",
      "Feb 01 17:50:00 ..... finished mapping\n",
      "Feb 01 17:50:10 ..... started Solo counting\n",
      "Feb 01 18:03:35 ..... finished Solo counting\n",
      "Feb 01 18:03:35 ..... started sorting BAM\n",
      "Feb 01 18:12:03 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "dir=demultiplexed_dropseq_sub_whitelist\n",
    "mkdir $dir\n",
    "for sample in fastq_dropseq_sub/*SRR1853182_2.SUB.fastq.gz\n",
    "do\n",
    "    NAME=${sample%_2.SUB.fastq.gz}\n",
    "    OUT=${NAME##*/}\n",
    "    echo $NAME\n",
    "    star_solo_hydrop_rna_mapping_unique_cellfilter_dropseq \\\n",
    "        ${NAME}_2.SUB.fastq.gz \\\n",
    "        ${NAME}_1.SUB.fastq.gz \\\n",
    "        $dir/$OUT.bam\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e36276a",
   "metadata": {
    "papermill": {
     "duration": 0.019128,
     "end_time": "2022-01-31T05:45:48.264739",
     "exception": false,
     "start_time": "2022-01-31T05:45:48.245611",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Now, we can get a summary of the statistics below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "14a6d071",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-01-31T05:45:48.304079Z",
     "iopub.status.busy": "2022-01-31T05:45:48.303650Z",
     "iopub.status.idle": "2022-01-31T05:45:48.903532Z",
     "shell.execute_reply": "2022-01-31T05:45:48.902474Z"
    },
    "papermill": {
     "duration": 0.623015,
     "end_time": "2022-01-31T05:45:48.903784",
     "exception": false,
     "start_time": "2022-01-31T05:45:48.280769",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR1853178.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,510210716\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.757257\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.62094\n",
      "Reads Mapped to Genome: Unique+Multiple,0.545704\n",
      "Reads Mapped to Genome: Unique,0.545704\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.372239\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.362322\n",
      "Estimated Number of Cells,2629\n",
      "Reads in Cells Mapped to Unique Genes,54743780\n",
      "Fraction of Reads in Cells,0.296135\n",
      "Mean Reads per Cell,20823\n",
      "Median Reads per Cell,15973\n",
      "UMIs in Cells,7773057\n",
      "Mean UMI per Cell,2956\n",
      "Median UMI per Cell,2202\n",
      "Mean Genes per Cell,1651\n",
      "Median Genes per Cell,1391\n",
      "Total Genes Detected,22048\n",
      "\n",
      "SRR1853179.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,524537760\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.768773\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.620952\n",
      "Reads Mapped to Genome: Unique+Multiple,0.543425\n",
      "Reads Mapped to Genome: Unique,0.543425\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.376026\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.365333\n",
      "Estimated Number of Cells,2208\n",
      "Reads in Cells Mapped to Unique Genes,57388174\n",
      "Fraction of Reads in Cells,0.299472\n",
      "Mean Reads per Cell,25991\n",
      "Median Reads per Cell,19780\n",
      "UMIs in Cells,7713453\n",
      "Mean UMI per Cell,3493\n",
      "Median UMI per Cell,2625\n",
      "Mean Genes per Cell,1863\n",
      "Median Genes per Cell,1603\n",
      "Total Genes Detected,22163\n",
      "\n",
      "SRR1853180.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,411365100\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.830703\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.566498\n",
      "Reads Mapped to Genome: Unique+Multiple,0.542113\n",
      "Reads Mapped to Genome: Unique,0.542113\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.363961\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.353768\n",
      "Estimated Number of Cells,1697\n",
      "Reads in Cells Mapped to Unique Genes,52238936\n",
      "Fraction of Reads in Cells,0.358962\n",
      "Mean Reads per Cell,30783\n",
      "Median Reads per Cell,22655\n",
      "UMIs in Cells,4566499\n",
      "Mean UMI per Cell,2690\n",
      "Median UMI per Cell,1953\n",
      "Mean Genes per Cell,1527\n",
      "Median Genes per Cell,1266\n",
      "Total Genes Detected,21708\n",
      "\n",
      "SRR1853181.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,486748074\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.674636\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.608495\n",
      "Reads Mapped to Genome: Unique+Multiple,0.332314\n",
      "Reads Mapped to Genome: Unique,0.332314\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.239943\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.23384\n",
      "Estimated Number of Cells,1798\n",
      "Reads in Cells Mapped to Unique Genes,38185717\n",
      "Fraction of Reads in Cells,0.335488\n",
      "Mean Reads per Cell,21237\n",
      "Median Reads per Cell,15512\n",
      "UMIs in Cells,8390763\n",
      "Mean UMI per Cell,4666\n",
      "Median UMI per Cell,3400\n",
      "Mean Genes per Cell,2211\n",
      "Median Genes per Cell,1907\n",
      "Total Genes Detected,22145\n",
      "\n",
      "SRR1853182.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,491102701\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.771998\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.637598\n",
      "Reads Mapped to Genome: Unique+Multiple,0.535425\n",
      "Reads Mapped to Genome: Unique,0.535425\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.382315\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.372148\n",
      "Estimated Number of Cells,2221\n",
      "Reads in Cells Mapped to Unique Genes,67329155\n",
      "Fraction of Reads in Cells,0.368397\n",
      "Mean Reads per Cell,30314\n",
      "Median Reads per Cell,23341\n",
      "UMIs in Cells,10082819\n",
      "Mean UMI per Cell,4539\n",
      "Median UMI per Cell,3363\n",
      "Mean Genes per Cell,2182\n",
      "Median Genes per Cell,1872\n",
      "Total Genes Detected,22631\n",
      "\n",
      "SRR1853183.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,438843919\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.844932\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.591365\n",
      "Reads Mapped to Genome: Unique+Multiple,0.551568\n",
      "Reads Mapped to Genome: Unique,0.551568\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.435267\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.423811\n",
      "Estimated Number of Cells,2249\n",
      "Reads in Cells Mapped to Unique Genes,79424580\n",
      "Fraction of Reads in Cells,0.427044\n",
      "Mean Reads per Cell,35315\n",
      "Median Reads per Cell,26822\n",
      "UMIs in Cells,6949914\n",
      "Mean UMI per Cell,3090\n",
      "Median UMI per Cell,2272\n",
      "Mean Genes per Cell,1716\n",
      "Median Genes per Cell,1447\n",
      "Total Genes Detected,21949\n",
      "\n",
      "SRR1853184.Solo.out/Gene/Summary.csv\n",
      "Number of Reads,494030802\n",
      "Reads With Valid Barcodes,1\n",
      "Sequencing Saturation,0.725467\n",
      "Q30 Bases in CB+UMI,-nan\n",
      "Q30 Bases in RNA read,0.600765\n",
      "Reads Mapped to Genome: Unique+Multiple,0.562359\n",
      "Reads Mapped to Genome: Unique,0.562359\n",
      "Reads Mapped to Transcriptome: Unique+Multipe Genes,0.416486\n",
      "Reads Mapped to Transcriptome: Unique Genes,0.405025\n",
      "Estimated Number of Cells,1953\n",
      "Reads in Cells Mapped to Unique Genes,45691499\n",
      "Fraction of Reads in Cells,0.228349\n",
      "Mean Reads per Cell,23395\n",
      "Median Reads per Cell,16929\n",
      "UMIs in Cells,8016444\n",
      "Mean UMI per Cell,4104\n",
      "Median UMI per Cell,2916\n",
      "Mean Genes per Cell,2005\n",
      "Median Genes per Cell,1688\n",
      "Total Genes Detected,21366\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for sample in demultiplexed_dropseq/*/Gene/Summary.csv\n",
    "do\n",
    "    echo ${sample#*/}\n",
    "    cat $sample\n",
    "    cut -d, -f2 $sample > ${sample%/Gene/Summary.csv}/Gene/Summary_numbers.csv\n",
    "    printf \"\\n\"\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
