{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f8b38fbf",
   "metadata": {},
   "source": [
    "Download these: https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/nuclei_2k  \n",
    "        https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/nuclei_900"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4aa41ed5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nuclei_2k_filtered_gene_bc_matrices\n",
      "nuclei_2k_filtered_gene_bc_matrices.tar.gz\n",
      "nuclei_2k_metrics_summary.csv\n",
      "nuclei_900_filtered_gene_bc_matrices\n",
      "nuclei_900_filtered_gene_bc_matrices.tar.gz\n",
      "nuclei_900_metrics_summary.csv\n"
     ]
    }
   ],
   "source": [
    "!ls 10x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c45addd5",
   "metadata": {
    "papermill": {
     "duration": 37.198864,
     "end_time": "2022-02-01T18:23:27.093168",
     "exception": false,
     "start_time": "2022-02-01T18:22:49.894304",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "44407da8",
   "metadata": {
    "papermill": {
     "duration": 15.648086,
     "end_time": "2022-02-01T18:23:42.741620",
     "exception": false,
     "start_time": "2022-02-01T18:23:27.093534",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.8.1 anndata==0.7.5 umap==0.5.1 numpy==1.20.2 scipy==1.6.2 pandas==1.3.3 scikit-learn==0.24.1 statsmodels==0.12.2 python-igraph==0.9.0 leidenalg==0.8.3 pynndescent==0.5.2\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=150, facecolor=\"white\", vector_friendly=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca25feba",
   "metadata": {
    "papermill": {
     "duration": 9.2e-05,
     "end_time": "2022-02-01T18:23:42.741858",
     "exception": false,
     "start_time": "2022-02-01T18:23:42.741766",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Make sure that the packages that you have match the ones below IF you want to verify our results and obtain exactly the same results as described below. If you want to simply analyse your own dataset, then most likely later versions will also be fine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8901d02b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "papermill": {
     "duration": 5.889459,
     "end_time": "2022-02-01T18:23:48.631445",
     "exception": false,
     "start_time": "2022-02-01T18:23:42.741986",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Package                            Version\n",
      "---------------------------------- -----------------\n",
      "adjustText                         0.7.3\n",
      "aiohttp                            3.7.4.post0\n",
      "aiohttp-cors                       0.7.0\n",
      "aioredis                           1.3.1\n",
      "alabaster                          0.7.12\n",
      "anaconda-client                    1.7.2\n",
      "anaconda-navigator                 1.9.7\n",
      "anaconda-project                   0.8.3\n",
      "anndata                            0.7.5\n",
      "annoy                              1.17.0\n",
      "ansiwrap                           0.8.4\n",
      "anyio                              2.2.0\n",
      "appdirs                            1.4.4\n",
      "arboreto                           0.1.6\n",
      "argcomplete                        1.12.2\n",
      "argh                               0.26.2\n",
      "argon2-cffi                        20.1.0\n",
      "asciitree                          0.3.3\n",
      "asn1crypto                         1.4.0\n",
      "astroid                            2.5\n",
      "astropy                            3.2.1\n",
      "async-timeout                      3.0.1\n",
      "atomicwrites                       1.4.0\n",
      "attrs                              20.3.0\n",
      "awscli                             1.22.48\n",
      "Babel                              2.9.0\n",
      "backcall                           0.2.0\n",
      "backports.functools-lru-cache      1.6.1\n",
      "backports.os                       0.1.1\n",
      "backports.shutil-get-terminal-size 1.0.0\n",
      "backports.tempfile                 1.0\n",
      "backports.weakref                  1.0.post1\n",
      "bap-atac                           0.6.7\n",
      "bash-kernel                        0.7.2\n",
      "bbknn                              1.4.0\n",
      "bcrypt                             3.2.0\n",
      "beautifulsoup4                     4.9.3\n",
      "bgzip                              0.3.5\n",
      "bhtsne                             0.1.9\n",
      "biopython                          1.77\n",
      "bitarray                           1.6.3\n",
      "bkcharts                           0.2\n",
      "black                              20.8b1\n",
      "bleach                             3.3.0\n",
      "blessings                          1.7\n",
      "blis                               0.7.4\n",
      "bokeh                              2.3.0\n",
      "boltons                            20.2.1\n",
      "boto                               2.49.0\n",
      "boto3                              1.20.48\n",
      "botocore                           1.23.48\n",
      "Bottleneck                         1.3.2\n",
      "Brotli                             1.0.9\n",
      "brotlipy                           0.7.0\n",
      "cached-property                    1.5.2\n",
      "cachetools                         4.2.1\n",
      "cairocffi                          1.3.0\n",
      "catalogue                          1.0.0\n",
      "certifi                            2021.10.8\n",
      "cffi                               1.14.5\n",
      "chardet                            4.0.0\n",
      "click                              7.1.2\n",
      "cloudpickle                        1.6.0\n",
      "clyent                             1.2.2\n",
      "colorama                           0.4.4\n",
      "colorful                           0.5.4\n",
      "conda                              4.11.0\n",
      "conda-build                        3.18.8\n",
      "conda-package-handling             1.7.2\n",
      "conda-verify                       3.4.2\n",
      "ConfigArgParse                     1.2.3\n",
      "contextlib2                        0.6.0.post1\n",
      "cooler                             0.8.11\n",
      "coverage                           6.3.1\n",
      "cryptography                       3.3.1\n",
      "cssselect2                         0.4.1\n",
      "ctxcore                            0.1.1\n",
      "cycler                             0.10.0\n",
      "cymem                              2.0.5\n",
      "Cython                             0.29.22\n",
      "cytoolz                            0.10.1\n",
      "dask                               2021.3.1\n",
      "datrie                             0.8.2\n",
      "decorator                          4.4.2\n",
      "deepTools                          3.5.1\n",
      "deeptoolsintervals                 0.1.9\n",
      "defusedxml                         0.6.0\n",
      "deprecation                        2.1.0\n",
      "dill                               0.3.4.dev0\n",
      "distributed                        2021.3.1\n",
      "docutils                           0.16\n",
      "ecdsa                              0.18.0b2\n",
      "entrypoints                        0.3\n",
      "et-xmlfile                         1.0.1\n",
      "fastcache                          1.1.0\n",
      "fastcluster                        1.2.4\n",
      "fastkde                            1.0.19\n",
      "fbpca                              1.0\n",
      "filelock                           3.0.12\n",
      "Flask                              1.1.2\n",
      "fonttools                          4.29.1\n",
      "frozendict                         1.2\n",
      "fsspec                             0.8.7\n",
      "future                             0.18.2\n",
      "fuzzysearch                        0.7.3\n",
      "gensim                             3.8.3\n",
      "geosketch                          1.2\n",
      "germalemma                         0.1.3\n",
      "get-version                        2.1\n",
      "gevent                             21.1.1\n",
      "gitdb                              4.0.5\n",
      "GitPython                          3.1.3\n",
      "glob2                              0.7\n",
      "globre                             0.1.5\n",
      "gmpy2                              2.0.8\n",
      "google-api-core                    1.26.1\n",
      "google-auth                        1.27.1\n",
      "googleapis-common-protos           1.53.0\n",
      "gpustat                            0.6.0\n",
      "greenlet                           1.0.0\n",
      "grpcio                             1.36.1\n",
      "gspread                            3.1.0\n",
      "h5py                               3.2.1\n",
      "harmonypy                          0.0.5\n",
      "HeapDict                           1.0.1\n",
      "hiredis                            1.1.0\n",
      "html5lib                           1.1\n",
      "httplib2                           0.17.0\n",
      "idna                               2.10\n",
      "imageio                            2.9.0\n",
      "imagesize                          1.2.0\n",
      "importlib-metadata                 3.10.0\n",
      "intel-openmp                       2020.0.133\n",
      "interlap                           0.2.7\n",
      "intervaltree                       2.1.0\n",
      "ipykernel                          5.5.0\n",
      "ipympl                             0.7.0\n",
      "ipython                            7.21.0\n",
      "ipython-genutils                   0.2.0\n",
      "ipywidgets                         7.6.3\n",
      "isort                              5.7.0\n",
      "itsdangerous                       1.1.0\n",
      "jdcal                              1.4.1\n",
      "jedi                               0.18.0\n",
      "jeepney                            0.6.0\n",
      "jgraph                             0.2.1\n",
      "Jinja2                             2.11.3\n",
      "jmespath                           0.10.0\n",
      "joblib                             1.0.1\n",
      "json5                              0.9.5\n",
      "jsonschema                         3.0.1\n",
      "jupyter                            1.0.0\n",
      "jupyter-client                     6.1.12\n",
      "jupyter-console                    6.2.0\n",
      "jupyter-contrib-core               0.3.3\n",
      "jupyter-contrib-nbextensions       0.5.1\n",
      "jupyter-core                       4.7.1\n",
      "jupyter-highlight-selected-word    0.2.0\n",
      "jupyter-latex-envs                 1.4.6\n",
      "jupyter-nbextensions-configurator  0.4.1\n",
      "jupyter-packaging                  0.7.12\n",
      "jupyter-server                     1.5.1\n",
      "jupyterlab                         3.0.12\n",
      "jupyterlab-code-formatter          1.4.5\n",
      "jupyterlab-server                  2.3.0\n",
      "jupyterlab-widgets                 1.0.0\n",
      "keyring                            22.2.0\n",
      "kiwisolver                         1.3.1\n",
      "lazy-object-proxy                  1.5.2\n",
      "lda                                2.0.0\n",
      "legacy-api-wrap                    1.2\n",
      "leidenalg                          0.8.3\n",
      "libarchive-c                       2.9\n",
      "llvmlite                           0.36.0\n",
      "locket                             0.2.1\n",
      "loompy                             3.0.6\n",
      "loomxpy                            0.1.0\n",
      "lxml                               4.6.2\n",
      "MACS2                              2.2.6\n",
      "MarkupSafe                         1.1.1\n",
      "matplotlib                         3.3.4\n",
      "mccabe                             0.6.1\n",
      "mistune                            0.8.4\n",
      "mkl                                2019.0\n",
      "mkl-fft                            1.3.0\n",
      "mkl-random                         1.1.1\n",
      "mkl-service                        2.3.0\n",
      "mock                               4.0.3\n",
      "more-itertools                     8.7.0\n",
      "mpmath                             1.1.0\n",
      "msgpack                            1.0.2\n",
      "multidict                          5.1.0\n",
      "multipledispatch                   0.6.0\n",
      "multiprocess                       0.70.11.1\n",
      "multiprocessing-on-dill            3.5.0a4\n",
      "murmurhash                         1.0.5\n",
      "mypy-extensions                    0.4.3\n",
      "natsort                            7.1.1\n",
      "navigator-updater                  0.2.1\n",
      "nbclassic                          0.2.6\n",
      "nbclient                           0.5.4\n",
      "nbconvert                          5.5.0\n",
      "nbformat                           5.1.3\n",
      "ncls                               0.0.57\n",
      "nest-asyncio                       1.5.1\n",
      "networkx                           2.5\n",
      "nltk                               3.5\n",
      "nose                               1.3.7\n",
      "nose2                              0.10.0\n",
      "notebook                           6.0.0\n",
      "numba                              0.53.1\n",
      "numexpr                            2.7.2\n",
      "numpy                              1.20.2\n",
      "numpy-groupies                     0.9.13\n",
      "numpydoc                           1.1.0\n",
      "nvidia-ml-py3                      7.352.0\n",
      "oauth2client                       4.1.3\n",
      "olefile                            0.46\n",
      "opencensus                         0.7.12\n",
      "opencensus-context                 0.1.2\n",
      "openpyxl                           3.0.6\n",
      "optparse-pretty                    0.1.1\n",
      "packaging                          20.9\n",
      "pandas                             1.3.3\n",
      "pandocfilters                      1.4.3\n",
      "papermill                          2.3.3\n",
      "paramiko                           2.9.2\n",
      "parso                              0.8.1\n",
      "partd                              1.1.0\n",
      "path                               15.1.2\n",
      "pathlib2                           2.3.5\n",
      "pathos                             0.2.8.dev0\n",
      "pathspec                           0.8.1\n",
      "patsy                              0.5.1\n",
      "PatternLite                        3.6\n",
      "pbr                                3.1.1\n",
      "pep8                               1.7.1\n",
      "pexpect                            4.8.0\n",
      "PhenoGraph                         1.5.2\n",
      "pickleshare                        0.7.5\n",
      "Pillow                             8.1.2\n",
      "pip                                21.0.1\n",
      "pkginfo                            1.7.0\n",
      "plac                               1.1.3\n",
      "plotly                             4.14.3\n",
      "pluggy                             0.13.1\n",
      "ply                                3.11\n",
      "pox                                0.2.9\n",
      "ppft                               1.6.6.3\n",
      "preshed                            3.0.5\n",
      "prometheus-client                  0.9.0\n",
      "prompt-toolkit                     3.0.17\n",
      "protobuf                           3.15.5\n",
      "psutil                             5.8.0\n",
      "ptyprocess                         0.7.0\n",
      "py                                 1.10.0\n",
      "py-spy                             0.3.4\n",
      "py2bit                             0.3.0\n",
      "pyarrow                            0.16.0\n",
      "pyasn1                             0.4.8\n",
      "pyasn1-modules                     0.2.8\n",
      "pybedtools                         0.8.1\n",
      "pyBigWig                           0.3.18\n",
      "pybiomart                          0.2.0\n",
      "pycodestyle                        2.6.0\n",
      "pycosat                            0.6.3\n",
      "pycparser                          2.20\n",
      "pycrypto                           2.6.1\n",
      "pycurl                             7.43.0.6\n",
      "pydyf                              0.1.2\n",
      "pyfaidx                            0.5.9.5\n",
      "pyfasta                            0.5.2\n",
      "pyflakes                           2.2.0\n",
      "Pygments                           2.8.0\n",
      "pylint                             2.7.1\n",
      "PyNaCl                             1.5.0\n",
      "pynndescent                        0.5.2\n",
      "pyodbc                             4.0.0-unsupported\n",
      "pyOpenSSL                          20.0.1\n",
      "pypairix                           0.3.7\n",
      "pyparsing                          2.4.7\n",
      "Pyphen                             0.10.0\n",
      "pyranges                           0.0.95\n",
      "pyrle                              0.0.32\n",
      "pyrsistent                         0.17.3\n",
      "pysam                              0.15.3\n",
      "pyscenic                           0.11.2\n",
      "PySocks                            1.7.1\n",
      "pytest                             5.0.1\n",
      "pytest-arraydiff                   0.3\n",
      "pytest-astropy                     0.5.0\n",
      "pytest-doctestplus                 0.3.0\n",
      "pytest-openfiles                   0.3.2\n",
      "pytest-remotedata                  0.3.1\n",
      "python-dateutil                    2.8.1\n",
      "python-igraph                      0.9.0\n",
      "python-Levenshtein                 0.12.2\n",
      "python-louvain                     0.15\n",
      "python-oauth2                      1.1.1\n",
      "pytz                               2021.1\n",
      "PyWavelets                         1.1.1\n",
      "PyYAML                             5.4.1\n",
      "pyzmq                              22.0.3\n",
      "QtAwesome                          1.0.1\n",
      "qtconsole                          5.0.2\n",
      "QtPy                               1.9.0\n",
      "ratelimiter                        1.2.0.post0\n",
      "ray                                1.2.0\n",
      "redis                              3.5.3\n",
      "regex                              2020.11.13\n",
      "remote-ikernel                     0.4.6\n",
      "requests                           2.25.1\n",
      "requests-cache                     0.5.2\n",
      "retrying                           1.3.3\n",
      "rope                               0.18.0\n",
      "rsa                                4.7.2\n",
      "ruamel-yaml                        0.15.87\n",
      "ruamel.yaml                        0.16.10\n",
      "ruamel.yaml.clib                   0.2.0\n",
      "s3transfer                         0.5.1\n",
      "scanorama                          1.7.1\n",
      "scanpy                             1.8.1\n",
      "scikit-image                       0.17.2\n",
      "scikit-learn                       0.24.1\n",
      "scipy                              1.6.2\n",
      "scrublet                           0.2.3\n",
      "seaborn                            0.11.1\n",
      "SecretStorage                      3.3.1\n",
      "Send2Trash                         1.5.0\n",
      "seqc                               0.2.0\n",
      "setuptools                         54.2.0\n",
      "simplegeneric                      0.8.1\n",
      "simplejson                         3.17.2\n",
      "sinfo                              0.3.1\n",
      "singledispatch                     0.0.0\n",
      "sinto                              0.7.1\n",
      "six                                1.15.0\n",
      "sklearn                            0.0\n",
      "smart-open                         4.2.0\n",
      "smmap                              3.0.4\n",
      "snakemake                          5.19.3\n",
      "snaptools                          1.4.8\n",
      "sniffio                            1.2.0\n",
      "snowballstemmer                    2.1.0\n",
      "sorted-nearest                     0.0.32\n",
      "sortedcollections                  2.1.0\n",
      "sortedcontainers                   2.3.0\n",
      "soupsieve                          2.2\n",
      "spacy                              2.3.5\n",
      "Sphinx                             3.5.1\n",
      "sphinxcontrib-applehelp            1.0.2\n",
      "sphinxcontrib-devhelp              1.0.2\n",
      "sphinxcontrib-htmlhelp             1.0.3\n",
      "sphinxcontrib-jsmath               1.0.1\n",
      "sphinxcontrib-qthelp               1.0.3\n",
      "sphinxcontrib-serializinghtml      1.1.4\n",
      "sphinxcontrib-websupport           1.2.4\n",
      "spyder                             3.3.6\n",
      "spyder-kernels                     0.5.1\n",
      "SQLAlchemy                         1.3.23\n",
      "srsly                              1.0.5\n",
      "statsmodels                        0.12.2\n",
      "stdlib-list                        0.8.0\n",
      "subsample                          0.0.6\n",
      "sympy                              1.7.1\n",
      "tables                             3.6.1\n",
      "tabulate                           0.8.9\n",
      "tblib                              1.7.0\n",
      "tenacity                           8.0.1\n",
      "terminado                          0.9.2\n",
      "testpath                           0.4.4\n",
      "texttable                          1.6.3\n",
      "textwrap3                          0.9.2\n",
      "thinc                              7.4.5\n",
      "threadpoolctl                      2.1.0\n",
      "tifffile                           2020.10.1\n",
      "tinycss2                           1.1.1\n",
      "tinydb                             4.6.1\n",
      "tmtoolkit                          0.9.0\n",
      "toml                               0.10.1\n",
      "toolz                              0.11.1\n",
      "toposort                           1.5\n",
      "tornado                            6.1\n",
      "tqdm                               4.59.0\n",
      "traitlets                          5.0.5\n",
      "typed-ast                          1.4.2\n",
      "typing                             3.7.4.3\n",
      "typing-extensions                  3.7.4.3\n",
      "umap                               0.1.1\n",
      "umap-learn                         0.5.1\n",
      "uncertainties                      3.1.2\n",
      "unicodecsv                         0.14.1\n",
      "urllib3                            1.26.4\n",
      "wasabi                             0.8.2\n",
      "wcwidth                            0.2.5\n",
      "weasyprint                         54.1\n",
      "webencodings                       0.5.1\n",
      "Werkzeug                           1.0.1\n",
      "wheel                              0.36.2\n",
      "widgetsnbextension                 3.5.1\n",
      "wikipedia                          1.4.0\n",
      "wrapt                              1.12.1\n",
      "wurlitzer                          2.0.1\n",
      "xlrd                               2.0.1\n",
      "XlsxWriter                         1.3.7\n",
      "xlwt                               1.3.0\n",
      "yarl                               1.6.3\n",
      "zict                               2.0.0\n",
      "zipp                               3.4.1\n",
      "zope.event                         4.5.0\n",
      "zope.interface                     5.2.0\n",
      "zopfli                             0.1.9\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce53de4a",
   "metadata": {
    "papermill": {
     "duration": 0.000117,
     "end_time": "2022-02-01T18:23:48.631897",
     "exception": false,
     "start_time": "2022-02-01T18:23:48.631780",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# drop-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5206a991",
   "metadata": {
    "papermill": {
     "duration": 0.005167,
     "end_time": "2022-02-01T18:23:48.637151",
     "exception": false,
     "start_time": "2022-02-01T18:23:48.631984",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ff62ca00",
   "metadata": {
    "papermill": {
     "duration": 0.01237,
     "end_time": "2022-02-01T18:23:48.649622",
     "exception": false,
     "start_time": "2022-02-01T18:23:48.637252",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['10x/nuclei_2k_filtered_gene_bc_matrices/mm10/',\n",
       " '10x/nuclei_900_filtered_gene_bc_matrices/mm10/']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx_dirs = sorted(glob.glob('10x/*/mm10/'))\n",
    "mtx_dirs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d04883f3",
   "metadata": {
    "papermill": {
     "duration": 34.212622,
     "end_time": "2022-02-01T18:24:22.870772",
     "exception": true,
     "start_time": "2022-02-01T18:23:48.658150",
     "status": "failed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... reading from cache file cache/10x-nuclei_2k_filtered_gene_bc_matrices-mm10-matrix.h5ad\n",
      "... writing an h5ad cache file to speedup reading next time\n"
     ]
    }
   ],
   "source": [
    "for mtx_dir in mtx_dirs:\n",
    "    sample=mtx_dir.replace('10x/', '')\n",
    "    sample=sample.replace('/mm10/', '')\n",
    "\n",
    "    adata = sc.read_10x_mtx(\n",
    "        mtx_dir,  # the directory with the `.mtx` file\n",
    "        var_names=\"gene_symbols\",  # use gene symbols for the variable names (variables-axis index)\n",
    "        cache=True,\n",
    "        )  # write a cache file for faster subsequent reading\n",
    "    adata.obs_names = sample + '_' + adata.obs_names\n",
    "    adata.var_names_make_unique()\n",
    "    sc.pp.calculate_qc_metrics(adata, inplace=True)\n",
    "    adata.obs.sort_values(by='n_genes_by_counts', ascending=False, inplace=True)\n",
    "    filename = 'tables/10x_' + sample + '.tsv'\n",
    "    if not os.path.exists(filename):\n",
    "        adata.obs.to_csv(filename, sep='\\t', header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4ff9658f",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = 'tables/10x_nuclei_2k_filtered_gene_bc_matrices.tsv'\n",
    "df = pd.read_csv(file, sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8076e271",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATGCCACG-1</th>\n",
       "      <td>7635</td>\n",
       "      <td>8.940629</td>\n",
       "      <td>25324.0</td>\n",
       "      <td>10.139547</td>\n",
       "      <td>17.370873</td>\n",
       "      <td>21.213079</td>\n",
       "      <td>26.658506</td>\n",
       "      <td>37.395356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGCTGTTA-1</th>\n",
       "      <td>4255</td>\n",
       "      <td>8.356085</td>\n",
       "      <td>8921.0</td>\n",
       "      <td>9.096275</td>\n",
       "      <td>18.731084</td>\n",
       "      <td>23.136420</td>\n",
       "      <td>29.402533</td>\n",
       "      <td>42.484026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGGAAACG-1</th>\n",
       "      <td>3221</td>\n",
       "      <td>8.077758</td>\n",
       "      <td>5680.0</td>\n",
       "      <td>8.644882</td>\n",
       "      <td>20.809859</td>\n",
       "      <td>25.228873</td>\n",
       "      <td>31.373239</td>\n",
       "      <td>44.559859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCAAACTG-1</th>\n",
       "      <td>2735</td>\n",
       "      <td>7.914252</td>\n",
       "      <td>6916.0</td>\n",
       "      <td>8.841738</td>\n",
       "      <td>46.515327</td>\n",
       "      <td>49.768652</td>\n",
       "      <td>54.525737</td>\n",
       "      <td>64.213418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTAACTGG-1</th>\n",
       "      <td>2654</td>\n",
       "      <td>7.884200</td>\n",
       "      <td>5382.0</td>\n",
       "      <td>8.591002</td>\n",
       "      <td>29.059829</td>\n",
       "      <td>34.039391</td>\n",
       "      <td>41.062802</td>\n",
       "      <td>54.589372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_CCTTCCCAGACAATAC-1</th>\n",
       "      <td>43</td>\n",
       "      <td>3.784190</td>\n",
       "      <td>407.0</td>\n",
       "      <td>6.011267</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_CTTGGCTGTCGGATCC-1</th>\n",
       "      <td>42</td>\n",
       "      <td>3.761200</td>\n",
       "      <td>397.0</td>\n",
       "      <td>5.986452</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_TTCTACATCCCTAACC-1</th>\n",
       "      <td>41</td>\n",
       "      <td>3.737670</td>\n",
       "      <td>402.0</td>\n",
       "      <td>5.998937</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_TCAGGATGTGCTAGCC-1</th>\n",
       "      <td>40</td>\n",
       "      <td>3.713572</td>\n",
       "      <td>531.0</td>\n",
       "      <td>6.276643</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_ATCATGGTCGTATCAG-1</th>\n",
       "      <td>39</td>\n",
       "      <td>3.688879</td>\n",
       "      <td>396.0</td>\n",
       "      <td>5.983936</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2352 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    n_genes_by_counts  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...               7635   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...               4255   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...               3221   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...               2735   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...               2654   \n",
       "...                                                               ...   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CCTTCCCAGAC...                 43   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CTTGGCTGTCG...                 42   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TTCTACATCCC...                 41   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TCAGGATGTGC...                 40   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATCATGGTCGT...                 39   \n",
       "\n",
       "                                                    log1p_n_genes_by_counts  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...                 8.940629   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...                 8.356085   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...                 8.077758   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...                 7.914252   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...                 7.884200   \n",
       "...                                                                     ...   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CCTTCCCAGAC...                 3.784190   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CTTGGCTGTCG...                 3.761200   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TTCTACATCCC...                 3.737670   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TCAGGATGTGC...                 3.713572   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATCATGGTCGT...                 3.688879   \n",
       "\n",
       "                                                    total_counts  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...       25324.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...        8921.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...        5680.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...        6916.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...        5382.0   \n",
       "...                                                          ...   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CCTTCCCAGAC...         407.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CTTGGCTGTCG...         397.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TTCTACATCCC...         402.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TCAGGATGTGC...         531.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATCATGGTCGT...         396.0   \n",
       "\n",
       "                                                    log1p_total_counts  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...           10.139547   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...            9.096275   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...            8.644882   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...            8.841738   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...            8.591002   \n",
       "...                                                                ...   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CCTTCCCAGAC...            6.011267   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CTTGGCTGTCG...            5.986452   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TTCTACATCCC...            5.998937   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TCAGGATGTGC...            6.276643   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATCATGGTCGT...            5.983936   \n",
       "\n",
       "                                                    pct_counts_in_top_50_genes  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...                   17.370873   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...                   18.731084   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...                   20.809859   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...                   46.515327   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...                   29.059829   \n",
       "...                                                                        ...   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CCTTCCCAGAC...                  100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CTTGGCTGTCG...                  100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TTCTACATCCC...                  100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TCAGGATGTGC...                  100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATCATGGTCGT...                  100.000000   \n",
       "\n",
       "                                                    pct_counts_in_top_100_genes  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...                    21.213079   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...                    23.136420   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...                    25.228873   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...                    49.768652   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...                    34.039391   \n",
       "...                                                                         ...   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CCTTCCCAGAC...                   100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CTTGGCTGTCG...                   100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TTCTACATCCC...                   100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TCAGGATGTGC...                   100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATCATGGTCGT...                   100.000000   \n",
       "\n",
       "                                                    pct_counts_in_top_200_genes  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...                    26.658506   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...                    29.402533   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...                    31.373239   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...                    54.525737   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...                    41.062802   \n",
       "...                                                                         ...   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CCTTCCCAGAC...                   100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_CTTGGCTGTCG...                   100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TTCTACATCCC...                   100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TCAGGATGTGC...                   100.000000   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATCATGGTCGT...                   100.000000   \n",
       "\n",
       "                                                    pct_counts_in_top_500_genes  \n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...                    37.395356  \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...                    42.484026  \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...                    44.559859  \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...                    64.213418  \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...                    54.589372  \n",
       "...                                                                         ...  \n",
       "nuclei_2k_filtered_gene_bc_matrices_CCTTCCCAGAC...                   100.000000  \n",
       "nuclei_2k_filtered_gene_bc_matrices_CTTGGCTGTCG...                   100.000000  \n",
       "nuclei_2k_filtered_gene_bc_matrices_TTCTACATCCC...                   100.000000  \n",
       "nuclei_2k_filtered_gene_bc_matrices_TCAGGATGTGC...                   100.000000  \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATCATGGTCGT...                   100.000000  \n",
       "\n",
       "[2352 rows x 8 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1a76168d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3014623.0"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['total_counts'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "913fc230",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10203353.0"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = 'tables/10x_nuclei_900_filtered_gene_bc_matrices.tsv'\n",
    "df = pd.read_csv(file, sep='\\t', index_col=0)\n",
    "df['total_counts'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "559497b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_GGTGCGTAGCTAAACA-1</th>\n",
       "      <td>7109</td>\n",
       "      <td>8.869258</td>\n",
       "      <td>55824.0</td>\n",
       "      <td>10.929977</td>\n",
       "      <td>23.876827</td>\n",
       "      <td>35.144382</td>\n",
       "      <td>46.449556</td>\n",
       "      <td>60.117512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_CGCTATCCAACACCTA-1</th>\n",
       "      <td>6887</td>\n",
       "      <td>8.837536</td>\n",
       "      <td>50523.0</td>\n",
       "      <td>10.830204</td>\n",
       "      <td>24.711518</td>\n",
       "      <td>36.872316</td>\n",
       "      <td>47.550621</td>\n",
       "      <td>60.483344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_TTGACTTGTGTGGCTC-1</th>\n",
       "      <td>6725</td>\n",
       "      <td>8.813736</td>\n",
       "      <td>42758.0</td>\n",
       "      <td>10.663335</td>\n",
       "      <td>24.825764</td>\n",
       "      <td>36.042378</td>\n",
       "      <td>46.723420</td>\n",
       "      <td>59.478928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_CAAGTTGAGGTCGGAT-1</th>\n",
       "      <td>6621</td>\n",
       "      <td>8.798153</td>\n",
       "      <td>46334.0</td>\n",
       "      <td>10.743653</td>\n",
       "      <td>22.875210</td>\n",
       "      <td>33.126861</td>\n",
       "      <td>45.012302</td>\n",
       "      <td>59.258860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_TGCCCATCACATCTTT-1</th>\n",
       "      <td>6567</td>\n",
       "      <td>8.789965</td>\n",
       "      <td>45220.0</td>\n",
       "      <td>10.719316</td>\n",
       "      <td>24.862893</td>\n",
       "      <td>37.450243</td>\n",
       "      <td>48.268465</td>\n",
       "      <td>61.393189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_TGCCCATAGACGCACA-1</th>\n",
       "      <td>130</td>\n",
       "      <td>4.875197</td>\n",
       "      <td>5722.0</td>\n",
       "      <td>8.652248</td>\n",
       "      <td>98.601887</td>\n",
       "      <td>99.475708</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_TTGAACGAGTACGATA-1</th>\n",
       "      <td>126</td>\n",
       "      <td>4.844187</td>\n",
       "      <td>5465.0</td>\n",
       "      <td>8.606302</td>\n",
       "      <td>98.609332</td>\n",
       "      <td>99.524245</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_AAACGGGAGTAGGTGC-1</th>\n",
       "      <td>113</td>\n",
       "      <td>4.736198</td>\n",
       "      <td>5123.0</td>\n",
       "      <td>8.541691</td>\n",
       "      <td>98.770252</td>\n",
       "      <td>99.746242</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_TACGGATCACGGTAAG-1</th>\n",
       "      <td>105</td>\n",
       "      <td>4.663439</td>\n",
       "      <td>4167.0</td>\n",
       "      <td>8.335192</td>\n",
       "      <td>98.680106</td>\n",
       "      <td>99.880010</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_CGCTATCAGCGATATA-1</th>\n",
       "      <td>101</td>\n",
       "      <td>4.624973</td>\n",
       "      <td>4024.0</td>\n",
       "      <td>8.300281</td>\n",
       "      <td>98.732604</td>\n",
       "      <td>99.975149</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>954 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    n_genes_by_counts  \\\n",
       "nuclei_900_filtered_gene_bc_matrices_GGTGCGTAGC...               7109   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCCAA...               6887   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGACTTGTG...               6725   \n",
       "nuclei_900_filtered_gene_bc_matrices_CAAGTTGAGG...               6621   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATCAC...               6567   \n",
       "...                                                               ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                130   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                126   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                113   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                105   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                101   \n",
       "\n",
       "                                                    log1p_n_genes_by_counts  \\\n",
       "nuclei_900_filtered_gene_bc_matrices_GGTGCGTAGC...                 8.869258   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCCAA...                 8.837536   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGACTTGTG...                 8.813736   \n",
       "nuclei_900_filtered_gene_bc_matrices_CAAGTTGAGG...                 8.798153   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATCAC...                 8.789965   \n",
       "...                                                                     ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                 4.875197   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                 4.844187   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                 4.736198   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                 4.663439   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                 4.624973   \n",
       "\n",
       "                                                    total_counts  \\\n",
       "nuclei_900_filtered_gene_bc_matrices_GGTGCGTAGC...       55824.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCCAA...       50523.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGACTTGTG...       42758.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_CAAGTTGAGG...       46334.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATCAC...       45220.0   \n",
       "...                                                          ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...        5722.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...        5465.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...        5123.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...        4167.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...        4024.0   \n",
       "\n",
       "                                                    log1p_total_counts  \\\n",
       "nuclei_900_filtered_gene_bc_matrices_GGTGCGTAGC...           10.929977   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCCAA...           10.830204   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGACTTGTG...           10.663335   \n",
       "nuclei_900_filtered_gene_bc_matrices_CAAGTTGAGG...           10.743653   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATCAC...           10.719316   \n",
       "...                                                                ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...            8.652248   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...            8.606302   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...            8.541691   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...            8.335192   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...            8.300281   \n",
       "\n",
       "                                                    pct_counts_in_top_50_genes  \\\n",
       "nuclei_900_filtered_gene_bc_matrices_GGTGCGTAGC...                   23.876827   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCCAA...                   24.711518   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGACTTGTG...                   24.825764   \n",
       "nuclei_900_filtered_gene_bc_matrices_CAAGTTGAGG...                   22.875210   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATCAC...                   24.862893   \n",
       "...                                                                        ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                   98.601887   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                   98.609332   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                   98.770252   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                   98.680106   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                   98.732604   \n",
       "\n",
       "                                                    pct_counts_in_top_100_genes  \\\n",
       "nuclei_900_filtered_gene_bc_matrices_GGTGCGTAGC...                    35.144382   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCCAA...                    36.872316   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGACTTGTG...                    36.042378   \n",
       "nuclei_900_filtered_gene_bc_matrices_CAAGTTGAGG...                    33.126861   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATCAC...                    37.450243   \n",
       "...                                                                         ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                    99.475708   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                    99.524245   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                    99.746242   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                    99.880010   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                    99.975149   \n",
       "\n",
       "                                                    pct_counts_in_top_200_genes  \\\n",
       "nuclei_900_filtered_gene_bc_matrices_GGTGCGTAGC...                    46.449556   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCCAA...                    47.550621   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGACTTGTG...                    46.723420   \n",
       "nuclei_900_filtered_gene_bc_matrices_CAAGTTGAGG...                    45.012302   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATCAC...                    48.268465   \n",
       "...                                                                         ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                   100.000000   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                   100.000000   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                   100.000000   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                   100.000000   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                   100.000000   \n",
       "\n",
       "                                                    pct_counts_in_top_500_genes  \n",
       "nuclei_900_filtered_gene_bc_matrices_GGTGCGTAGC...                    60.117512  \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCCAA...                    60.483344  \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGACTTGTG...                    59.478928  \n",
       "nuclei_900_filtered_gene_bc_matrices_CAAGTTGAGG...                    59.258860  \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATCAC...                    61.393189  \n",
       "...                                                                         ...  \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                   100.000000  \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                   100.000000  \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                   100.000000  \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                   100.000000  \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                   100.000000  \n",
       "\n",
       "[954 rows x 8 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1df89db",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "source": [
    "Merge into one nice dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2780bea2",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tables/10x_nuclei_2k_filtered_gene_bc_matrices.tsv\n",
      "tables/10x_nuclei_900_filtered_gene_bc_matrices.tsv\n"
     ]
    }
   ],
   "source": [
    "df_merged = pd.DataFrame()\n",
    "for file in glob.glob('tables/10x_*.tsv'):\n",
    "    print(file)\n",
    "    df = pd.read_csv(file, sep='\\t', index_col=0)\n",
    "    df_merged = df_merged.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7b1b1777",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATGCCACG-1</th>\n",
       "      <td>7635</td>\n",
       "      <td>8.940629</td>\n",
       "      <td>25324.0</td>\n",
       "      <td>10.139547</td>\n",
       "      <td>17.370873</td>\n",
       "      <td>21.213079</td>\n",
       "      <td>26.658506</td>\n",
       "      <td>37.395356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGCTGTTA-1</th>\n",
       "      <td>4255</td>\n",
       "      <td>8.356085</td>\n",
       "      <td>8921.0</td>\n",
       "      <td>9.096275</td>\n",
       "      <td>18.731084</td>\n",
       "      <td>23.136420</td>\n",
       "      <td>29.402533</td>\n",
       "      <td>42.484026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGGAAACG-1</th>\n",
       "      <td>3221</td>\n",
       "      <td>8.077758</td>\n",
       "      <td>5680.0</td>\n",
       "      <td>8.644882</td>\n",
       "      <td>20.809859</td>\n",
       "      <td>25.228873</td>\n",
       "      <td>31.373239</td>\n",
       "      <td>44.559859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCAAACTG-1</th>\n",
       "      <td>2735</td>\n",
       "      <td>7.914252</td>\n",
       "      <td>6916.0</td>\n",
       "      <td>8.841738</td>\n",
       "      <td>46.515327</td>\n",
       "      <td>49.768652</td>\n",
       "      <td>54.525737</td>\n",
       "      <td>64.213418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTAACTGG-1</th>\n",
       "      <td>2654</td>\n",
       "      <td>7.884200</td>\n",
       "      <td>5382.0</td>\n",
       "      <td>8.591002</td>\n",
       "      <td>29.059829</td>\n",
       "      <td>34.039391</td>\n",
       "      <td>41.062802</td>\n",
       "      <td>54.589372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_TGCCCATAGACGCACA-1</th>\n",
       "      <td>130</td>\n",
       "      <td>4.875197</td>\n",
       "      <td>5722.0</td>\n",
       "      <td>8.652248</td>\n",
       "      <td>98.601887</td>\n",
       "      <td>99.475708</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_TTGAACGAGTACGATA-1</th>\n",
       "      <td>126</td>\n",
       "      <td>4.844187</td>\n",
       "      <td>5465.0</td>\n",
       "      <td>8.606302</td>\n",
       "      <td>98.609332</td>\n",
       "      <td>99.524245</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_AAACGGGAGTAGGTGC-1</th>\n",
       "      <td>113</td>\n",
       "      <td>4.736198</td>\n",
       "      <td>5123.0</td>\n",
       "      <td>8.541691</td>\n",
       "      <td>98.770252</td>\n",
       "      <td>99.746242</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_TACGGATCACGGTAAG-1</th>\n",
       "      <td>105</td>\n",
       "      <td>4.663439</td>\n",
       "      <td>4167.0</td>\n",
       "      <td>8.335192</td>\n",
       "      <td>98.680106</td>\n",
       "      <td>99.880010</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclei_900_filtered_gene_bc_matrices_CGCTATCAGCGATATA-1</th>\n",
       "      <td>101</td>\n",
       "      <td>4.624973</td>\n",
       "      <td>4024.0</td>\n",
       "      <td>8.300281</td>\n",
       "      <td>98.732604</td>\n",
       "      <td>99.975149</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3306 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    n_genes_by_counts  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...               7635   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...               4255   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...               3221   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...               2735   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...               2654   \n",
       "...                                                               ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                130   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                126   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                113   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                105   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                101   \n",
       "\n",
       "                                                    log1p_n_genes_by_counts  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...                 8.940629   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...                 8.356085   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...                 8.077758   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...                 7.914252   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...                 7.884200   \n",
       "...                                                                     ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                 4.875197   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                 4.844187   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                 4.736198   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                 4.663439   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                 4.624973   \n",
       "\n",
       "                                                    total_counts  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...       25324.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...        8921.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...        5680.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...        6916.0   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...        5382.0   \n",
       "...                                                          ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...        5722.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...        5465.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...        5123.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...        4167.0   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...        4024.0   \n",
       "\n",
       "                                                    log1p_total_counts  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...           10.139547   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...            9.096275   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...            8.644882   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...            8.841738   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...            8.591002   \n",
       "...                                                                ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...            8.652248   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...            8.606302   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...            8.541691   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...            8.335192   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...            8.300281   \n",
       "\n",
       "                                                    pct_counts_in_top_50_genes  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...                   17.370873   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...                   18.731084   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...                   20.809859   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...                   46.515327   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...                   29.059829   \n",
       "...                                                                        ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                   98.601887   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                   98.609332   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                   98.770252   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                   98.680106   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                   98.732604   \n",
       "\n",
       "                                                    pct_counts_in_top_100_genes  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...                    21.213079   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...                    23.136420   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...                    25.228873   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...                    49.768652   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...                    34.039391   \n",
       "...                                                                         ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                    99.475708   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                    99.524245   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                    99.746242   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                    99.880010   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                    99.975149   \n",
       "\n",
       "                                                    pct_counts_in_top_200_genes  \\\n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...                    26.658506   \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...                    29.402533   \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...                    31.373239   \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...                    54.525737   \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...                    41.062802   \n",
       "...                                                                         ...   \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                   100.000000   \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                   100.000000   \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                   100.000000   \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                   100.000000   \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                   100.000000   \n",
       "\n",
       "                                                    pct_counts_in_top_500_genes  \n",
       "nuclei_2k_filtered_gene_bc_matrices_TCTATTGCATG...                    37.395356  \n",
       "nuclei_2k_filtered_gene_bc_matrices_TACGGTACAGC...                    42.484026  \n",
       "nuclei_2k_filtered_gene_bc_matrices_AAACGGGTCGG...                    44.559859  \n",
       "nuclei_2k_filtered_gene_bc_matrices_ATAACGCTCCA...                    64.213418  \n",
       "nuclei_2k_filtered_gene_bc_matrices_GGAAAGCTCTA...                    54.589372  \n",
       "...                                                                         ...  \n",
       "nuclei_900_filtered_gene_bc_matrices_TGCCCATAGA...                   100.000000  \n",
       "nuclei_900_filtered_gene_bc_matrices_TTGAACGAGT...                   100.000000  \n",
       "nuclei_900_filtered_gene_bc_matrices_AAACGGGAGT...                   100.000000  \n",
       "nuclei_900_filtered_gene_bc_matrices_TACGGATCAC...                   100.000000  \n",
       "nuclei_900_filtered_gene_bc_matrices_CGCTATCAGC...                   100.000000  \n",
       "\n",
       "[3306 rows x 8 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
